// @kousum/semi-ui-vue/dist/banner/index.js
import { defineComponent as B, useSlots as M, reactive as N, onMounted as S, onUnmounted as j, createVNode as n, mergeProps as z, isVNode as _ } from "vue";
import g from "classnames";
import * as r from "../PropTypes";
import { vuePropsMake as k } from "../PropTypes";
import A from "../iconButton";
import { cssClasses as V, strings as E } from '@douyinfe/semi-foundation/lib/es/banner/constants';
import F from '@douyinfe/semi-foundation/lib/es/banner/foundation';
import '@douyinfe/semi-foundation/lib/es/banner/banner.css';
import { Title as O, Paragraph as X } from "../typography/index";
import { IconClose as D, IconAlertTriangle as J, IconTickCircle as R, IconInfoCircle as U, IconAlertCircle as Y } from "@kousum/semi-icons-vue";
import q from '@douyinfe/semi-foundation/lib/es/utils/warning';
import { useBaseComponent as G } from "../_base/baseComponent";
function y(t) {
  return typeof t == "function" || Object.prototype.toString.call(t) === "[object Object]" && !_(t);
}
const o = V.PREFIX;
E.TYPE;
const H = {
  // target: PropTypes.func,
  fullMode: r.bool,
  // insertAfter: PropTypes.func,
  type: String,
  title: r.node,
  description: r.node,
  icon: r.node,
  closeIcon: r.node,
  style: r.object,
  className: r.string,
  onClose: r.func,
  bordered: r.bool
}, K = {
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  onClose: () => {
  },
  type: "info",
  fullMode: !0
}, L = k(H, K), ce = /* @__PURE__ */ B({
  props: {
    ...L
  },
  name: "Banner",
  setup(t, {}) {
    q("target" in t, "[Semi Banner] 'target' has been deprecated, please write JSX directly instead.");
    const p = M(), d = N({
      visible: !0
    });
    let s;
    const {
      adapter: v,
      getDataAttr: C
    } = G(t, d);
    function $() {
      return {
        ...v(),
        setVisible: () => {
          d.visible = !1;
        },
        notifyClose: (e) => {
          const {
            onClose: l
          } = t;
          l(e);
        }
      };
    }
    const I = $();
    S(() => {
      s = new F(I), s.init();
    }), j(() => {
      s.destroy();
    });
    const P = (e) => {
      e && e.stopPropagation(), s.removeBanner(e);
    };
    function x() {
      const {
        closeIcon: e
      } = t;
      return e === null ? e : n(A, {
        className: `${o}-close`,
        onClick: P,
        icon: e || n(D, {
          "x-semi-prop": "closeIcon",
          "aria-hidden": !0
        }, null),
        theme: "borderless",
        size: "small",
        type: "tertiary",
        "aria-label": "Close"
      }, null);
    }
    function T() {
      const {
        type: e,
        icon: l
      } = t;
      let i = {
        warning: n(J, {
          size: "large",
          "aria-label": "warning"
        }, null),
        success: n(R, {
          size: "large",
          "aria-label": "success"
        }, null),
        info: n(U, {
          size: "large",
          "aria-label": "info"
        }, null),
        danger: n(Y, {
          size: "large",
          "aria-label": "danger"
        }, null)
      }[e];
      const u = g({
        [`${o}-icon`]: !0
        // [prefixCls + '-' + type]: true,
      });
      return typeof l < "u" && (i = l), i ? n("div", {
        class: u,
        "x-semi-prop": "icon"
      }, [i]) : null;
    }
    return () => {
      var b;
      const e = (b = p.default) == null ? void 0 : b.call(p), {
        type: l,
        className: f,
        style: i,
        bordered: u,
        title: a,
        description: c,
        fullMode: m
      } = t, {
        visible: h
      } = d, w = g(o, f, {
        [`${o}-${l}`]: l,
        [`${o}-full`]: m,
        [`${o}-in-container`]: !m,
        [`${o}-bordered`]: !m && u
      });
      return h ? n("div", z({
        class: w,
        style: i,
        role: "alert"
      }, C()), [n("div", {
        class: `${o}-content-wrapper`
      }, [n("div", {
        class: `${o}-content`
      }, [T(), n("div", {
        class: `${o}-content-body`
      }, [a ? n(O, {
        heading: 5,
        className: `${o}-title`,
        component_: "div",
        "x-semi-prop": "title"
      }, y(a) ? a : {
        default: () => [a]
      }) : null, c ? n(X, {
        className: `${o}-description`,
        component_: "div",
        "x-semi-prop": "description"
      }, y(c) ? c : {
        default: () => [c]
      }) : null])]), x()]), e ? n("div", {
        class: `${o}-extra`,
        "x-semi-prop": "children"
      }, [e]) : null]) : null;
    };
  }
});
export {
  ce as default,
  L as vuePropsType
};
