// @kousum/semi-ui-vue/dist/calendar/index.js
import { defineComponent as u, useSlots as i, createVNode as r, cloneVNode as d } from "vue";
import * as e from "../PropTypes";
import { vuePropsMake as c } from "../PropTypes";
import f from "./dayCalendar";
import y from "./weekCalendar";
import h from "./monthCalendar";
import T from "./rangeCalendar";
import '@douyinfe/semi-foundation/lib/es/calendar/calendar.css';
export * from "./interface";
const C = {
  displayValue: e.object,
  header: e.node,
  events: e.array,
  mode: e.string,
  showCurrTime: e.bool,
  weekStartsOn: e.number,
  scrollTop: e.number,
  onClick: e.func,
  renderTimeDisplay: e.func,
  renderDateDisplay: e.func,
  markWeekend: e.bool,
  minEventHeight: e.number,
  width: e.oneOfType([e.number, e.string]),
  height: e.oneOfType([e.number, e.string]),
  style: e.object,
  className: e.string,
  range: e.array,
  onClose: e.func,
  dateGridRender: e.func,
  allDayEventsRender: e.func
}, g = {
  events: [],
  displayValue: /* @__PURE__ */ new Date(),
  showCurrTime: !0,
  mode: "week",
  markWeekend: !1,
  height: 600,
  scrollTop: 400,
  weekStartsOn: 0
}, k = c(C, g), N = /* @__PURE__ */ u({
  props: {
    ...k
  },
  name: "Calendar",
  setup(o, {}) {
    return i(), () => {
      const {
        mode: n,
        range: t,
        showCurrTime: a,
        scrollTop: l,
        renderTimeDisplay: s,
        ...m
      } = o, p = {
        month: r(h, null, null),
        week: r(y, null, null),
        day: r(f, null, null),
        range: r(T, null, null)
      };
      return d(p[n], {
        ...m,
        // 去掉month不需要的参数
        ...n === "month" ? {} : {
          range: t,
          showCurrTime: a,
          scrollTop: l,
          renderTimeDisplay: s
        }
      });
    };
  }
});
export {
  N as default,
  k as vuePropsType
};
