// @kousum/semi-ui-vue/dist/calendar/timeCol.js
import { defineComponent as u, useSlots as f, createVNode as t } from "vue";
import y from "classnames";
import * as p from "../PropTypes";
import { vuePropsMake as T } from "../PropTypes";
import d from "../locale/localeConsumer";
import { cssClasses as C } from '@douyinfe/semi-foundation/lib/es/calendar/constants';
import '@douyinfe/semi-foundation/lib/es/calendar/calendar.css';
const i = `${C.PREFIX}-time`, $ = {
  className: p.string,
  renderTimeDisplay: p.func
}, P = T($, {}), A = /* @__PURE__ */ u({
  props: {
    ...P
  },
  name: "TimeCol",
  setup(m, {}) {
    f();
    function a(e) {
      const {
        renderTimeDisplay: n
      } = m;
      if (typeof n == "function")
        return n(e);
      {
        const o = (s, r) => s.replace("${time}", String(r));
        return t(d, {
          componentName: "Calendar",
          key: `locale-${e}`
        }, {
          default: (s) => {
            let r = e < 12 ? o(s.AM, e) : o(s.PM, e - 12);
            return e === 12 && (r = o(s.PM, e)), r;
          }
        });
      }
    }
    function c() {
      const {
        className: e
      } = m, n = y(e, `${i}`), o = [...Array(24).keys()].map((r) => a(r));
      o.splice(0, 1, "");
      const s = o.map((r, l) => t("li", {
        key: `time-${l}`,
        class: `${i}-item`
      }, [t("span", null, [r])]));
      return t("div", {
        class: n
      }, [t("ul", {
        class: `${i}-items`
      }, [s])]);
    }
    return () => c();
  }
});
export {
  A as default,
  P as vuePropsType
};
