// @kousum/semi-ui-vue/dist/toast/index.js
import { defineComponent as N, useSlots as W, shallowRef as B, getCurrentInstance as L, ref as X, reactive as _, createVNode as h, Fragment as q, mergeProps as j, createApp as H } from "vue";
import * as g from "../PropTypes";
import O from '@douyinfe/semi-foundation/lib/es/toast/toastListFoundation';
import { cssClasses as I, strings as U } from '@douyinfe/semi-foundation/lib/es/toast/constants';
import V from "./toast";
import '@douyinfe/semi-foundation/lib/es/toast/toast.css';
import w from '@douyinfe/semi-foundation/lib/es/utils/uuid';
import D from "./useToast";
import G from "../_cssAnimation";
import A from "classnames";
import { useBaseComponent as J } from "../_base/baseComponent";
const K = /* @__PURE__ */ N({
  props: {
    content: g.node,
    duration: g.number,
    onClose: g.func,
    icon: g.node,
    direction: String,
    stack: g.bool
    // ref: [Function, Object],
  },
  name: "ToastList",
  setup(p, {
    expose: m
  }) {
    W();
    const t = B(!1);
    function o(n) {
      t.value = n;
    }
    function s() {
      return t.value;
    }
    L();
    let f = X();
    const d = _({
      list: [],
      removedItems: [],
      updatedItems: [],
      mouseInSide: !1
    }), {
      adapter: C
    } = J(p, d);
    function T() {
      return {
        ...C(),
        updateToast: (n, l, v) => {
          d.list = n, d.removedItems = l, d.updatedItems = v;
        },
        handleMouseInSideChange: (n) => {
          d.mouseInSide = n;
        },
        getInputWrapperRect: () => {
          var n;
          return (n = f.value) == null ? void 0 : n.getBoundingClientRect();
        }
      };
    }
    const k = (n) => {
      t.value && c.handleMouseInSideChange(!0);
    }, x = (n) => {
      var l;
      t.value && ((l = c.getInputWrapperRect()) != null && l.height) && c.handleMouseInSideChange(!1);
    }, R = T(), c = new O(R);
    function y(n) {
      return c.hasToast(n);
    }
    function e(n) {
      return c.addToast(n);
    }
    function i(n, l) {
      return c.updateToast(n, l);
    }
    function u(n) {
      return c.removeToast(n);
    }
    function r() {
      return c.destroyAll();
    }
    return m({
      has: y,
      add: e,
      update: i,
      remove: u,
      destroyAll: r,
      setStack: o,
      getStack: s
    }), () => {
      let {
        list: n
      } = d;
      const {
        removedItems: l,
        updatedItems: v
      } = d;
      n = Array.from(/* @__PURE__ */ new Set([...n, ...l]));
      const E = v.map(({
        id: a
      }) => a), z = (a) => {
        var S;
        (S = a == null ? void 0 : a.foundation) != null && S._id && E.includes(a.foundation._id) && a.foundation.restartCloseTimer();
      };
      return h(q, null, [h("div", {
        class: A({
          [`${I.PREFIX}-innerWrapper`]: !0,
          [`${I.PREFIX}-innerWrapper-hover`]: d.mouseInSide
        }),
        ref: f,
        onMouseenter: k,
        onMouseleave: x
      }, [n.map((a, S) => {
        const P = l.find((b) => b.id === a.id) !== void 0;
        return h(G, {
          key: a.id,
          motion: a.motion,
          animationState: P ? "leave" : "enter",
          startClassName: P ? `${I.PREFIX}-animation-hide` : `${I.PREFIX}-animation-show`,
          children: ({
            animationClassName: b,
            animationEventsNeedBind: F,
            isAnimating: $
          }) => P && !$ ? null : h(V, j(a, {
            stack: t.value,
            stackExpanded: d.mouseInSide,
            positionInList: {
              length: n.length,
              index: S
            },
            className: A({
              [a.className]: !!a.className,
              [b]: !0
            })
          }, F, {
            style: {
              ...a.style
            },
            close: (M) => u(M),
            ref: z
          }), null)
        }, null);
      })])]);
    };
  }
});
function Q(p) {
  let m, t;
  const o = {
    motion: !0,
    zIndex: 1010,
    content: ""
  };
  let s;
  p && y(p);
  function f(e) {
    const i = e.id ?? w("toast");
    if (t) {
      const u = document.querySelector(`#${s}`);
      ["top", "left", "bottom", "right"].map((r) => {
        r in e && (u.style[r] = typeof e[r] == "number" ? `${e[r]}px` : e[r]);
      }), !!e.stack !== t.getStack() && t.setStack(!!e.stack), t.has(i) ? t.update(i, {
        ...e,
        id: i
      }) : t.add({
        ...e,
        id: i
      });
    } else {
      const u = document.createElement("div");
      s || (s = w("toast-wrapper").slice(0, 26)), u.className = I.WRAPPER, u.id = s, u.style.zIndex = String(typeof e.zIndex == "number" ? e.zIndex : o.zIndex), ["top", "left", "bottom", "right"].map((r) => {
        if (r in o || r in e) {
          const n = e[r] ? e[r] : o[r];
          u.style[r] = typeof n == "number" ? `${n}px` : n;
        }
      }), o.getPopupContainer ? o.getPopupContainer().appendChild(u) : document.body.appendChild(u), m = H(() => h(K, {
        ref: (r) => {
          t || (r.add({
            ...e,
            id: i
          }), r.setStack(!!e.stack)), t = r;
        }
      })), m.mount(u);
    }
    return i;
  }
  function d(e) {
    t && t.remove(e);
  }
  function C() {
    if (t) {
      t.destroyAll();
      const e = document.querySelector(`#${s}`);
      m.unmount(), e && e.parentNode.removeChild(e), t = null, s = null;
    }
  }
  function T() {
    return s;
  }
  function k(e) {
    return typeof e == "string" && (e = {
      content: e
    }), f({
      ...o,
      ...e,
      type: "info"
    });
  }
  function x(e) {
    return typeof e == "string" && (e = {
      content: e
    }), f({
      ...o,
      ...e,
      type: "warning"
    });
  }
  function R(e) {
    return typeof e == "string" && (e = {
      content: e
    }), f({
      ...o,
      ...e,
      type: "error"
    });
  }
  function c(e) {
    return typeof e == "string" && (e = {
      content: e
    }), f({
      ...o,
      ...e,
      type: "success"
    });
  }
  function y(e) {
    ["top", "left", "bottom", "right"].forEach((i) => {
      i in e && (o[i] = e[i]);
    }), typeof e.theme == "string" && U.themes.includes(e.theme) && (o.theme = e.theme), typeof e.zIndex == "number" && (o.zIndex = e.zIndex), typeof e.duration == "number" && (o.duration = e.duration), typeof e.getPopupContainer == "function" && (o.getPopupContainer = e.getPopupContainer);
  }
  return {
    create: f,
    close: d,
    destroyAll: C,
    getWrapperId: T,
    info: k,
    warning: x,
    error: R,
    success: c,
    config: y,
    useToast: D
  };
}
class Y {
  static create(m) {
    return Q(m);
  }
}
const ce = Y.create();
export {
  Y as ToastFactory,
  K as ToastList,
  ce as default,
  Q as useToastHook
};
