// @kousum/semi-ui-vue/dist/layout/index.js
import { defineComponent as l, h as x, createVNode as u, mergeProps as g, reactive as F, isVNode as H } from "vue";
import y from "classnames";
import { cssClasses as S } from '@douyinfe/semi-foundation/lib/es/layout/constants';
import '@douyinfe/semi-foundation/lib/es/layout/layout.css';
import P from "./layoutContext";
import v from "./Sider";
const B = {
  Header: "header",
  Footer: "footer",
  Content: "main",
  Layout: "section"
}, T = {
  prefixCls: {
    type: String,
    default: S.PREFIX
  },
  style: [String, Object],
  className: String,
  tagName: String,
  type: String
}, m = /* @__PURE__ */ l({
  props: {
    ...T
  },
  name: "Basic",
  setup(o, {
    slots: e
  }) {
    return () => {
      const {
        prefixCls: r,
        type: s,
        className: t,
        tagName: a,
        ...d
      } = o, i = y(t, `${r}-${s}`);
      return x(a, {
        className: i,
        ...d
      }, e.default ? e.default() : null);
    };
  }
});
function f(o) {
  const e = B[o], r = o.toLowerCase();
  return (s) => /* @__PURE__ */ l((t, {
    slots: a
  }) => () => u(s, g({
    role: t.role,
    "aria-label": t["aria-label"],
    type: r,
    tagName: e
  }, t), {
    default: () => [a.default ? a.default() : null]
  }));
}
const V = f("Header")(m), $ = f("Footer")(m), j = f("Content")(m), k = v, E = {
  prefixCls: {
    type: String,
    default: S.PREFIX
  },
  style: [String, Object],
  className: String,
  hasSider: Boolean,
  tagName: {
    type: String,
    default: "section"
  }
}, I = /* @__PURE__ */ l({
  props: {
    ...E
  },
  name: "Layout",
  setup(o, {
    slots: e
  }) {
    const r = F({
      siders: []
    });
    function s() {
      return {
        addSider: (t) => {
          r.siders = [...r.siders, t];
        },
        removeSider: (t) => {
          r.siders = r.siders.filter((a) => a !== t);
        }
      };
    }
    return () => {
      var p;
      const t = ((p = e.default) == null ? void 0 : p.call(e)) || [], {
        prefixCls: a,
        className: d,
        hasSider: i,
        tagName: N,
        ...C
      } = o, {
        siders: L
      } = r, h = y(d, a, {
        [`${a}-has-sider`]: typeof i == "boolean" && i || L.length > 0 || t.some((n) => H(n) && n.type && n.type.name === "Layout.Sider")
      }), b = N;
      return u(P.Provider, {
        value: {
          siderHook: s()
        }
      }, {
        default: () => [u(b, g({
          className: h
        }, C), {
          default: () => {
            var n;
            return [(n = e.default) == null ? void 0 : n.call(e)];
          }
        })]
      });
    };
  }
}), c = I;
c.Header = V;
c.Footer = $;
c.Content = j;
c.Sider = k;
export {
  j as LayoutContent,
  $ as LayoutFooter,
  V as LayoutHeader,
  k as LayoutSider,
  c as default,
  E as vuePropsType
};
