// @kousum/semi-ui-vue/dist/modal/__stories__/Demo.js
import { createVNode as o } from "vue";
import e from "../__test__/ModalDemo";
import a from "../__test__/ModalDemoHook";
import m from "../__test__/ModalDemoConfirm";
const u = {
  /* 👇 The title prop is optional.
   * See https://storybook.js.org/docs/7.0/vue/configure/overview#configure-story-loading
   * to learn how to generate automatic titles
   */
  title: "D展示类/Modal",
  render: (l) => ({
    setup() {
      return () => o("div", {
        style: {
          padding: "10px"
        }
      }, [l.name === "ModalDemo" ? o(e, null, null) : null, l.name === "ModalDemoHook" ? o(a, null, null) : null, l.name === "ModalDemoConfirm" ? o(m, null, null) : null]);
    }
  }),
  parameters: {
    // More on how to position stories at: https://storybook.js.org/docs/7.0/react/configure/story-layout
    layout: "fullscreen"
  },
  // This component will have an automatically generated docsPage entry: https://storybook.js.org/docs/7.0/vue/writing-docs/docs-page
  tags: ["autodocs"]
}, M = {
  args: {
    name: "ModalDemo"
  }
}, s = {
  args: {
    name: "ModalDemoHook"
  }
}, i = {
  args: {
    name: "ModalDemoConfirm"
  }
};
export {
  M as Modal,
  i as ModalConfirm,
  s as ModalHook,
  u as default
};
