// @kousum/semi-ui-vue/dist/table/ColumnFilter.js
import { defineComponent as X, ref as T, watch as N, isVNode as k, createVNode as p, cloneVNode as q, mergeProps as z } from "vue";
import H from "classnames";
import { pick as B, isEqual as G, noop as $ } from "lodash";
import { IconFilter as J } from "@kousum/semi-icons-vue";
import { cssClasses as M } from '@douyinfe/semi-foundation/lib/es/table/constants';
import K, { DropdownVuePropsType as E, DropdownMenu as L, DropdownItem as Q } from "../dropdown";
import { Radio as U } from "../radio";
import { Checkbox as W } from "../checkbox";
import * as a from "../PropTypes";
import { vuePropsMake as Y } from "../PropTypes";
import { useHasInProps as Z } from "../_base/baseComponent";
function A(e) {
  return typeof e == "function" || Object.prototype.toString.call(e) === "[object Object]" && !k(e);
}
function R(e, b = null, u = 0) {
  const {
    filterMultiple: w = !0,
    filters: F = [],
    filteredValue: y = [],
    filterDropdownVisible: h,
    onSelect: c = $,
    onFilterDropdownVisibleChange: l = $,
    trigger: x = "click",
    position: S = "bottom",
    renderFilterDropdown: g,
    renderFilterDropdownItem: C
  } = e ?? {}, t = B(e, ["tempFilteredValue", "setTempFilteredValue", "confirm", "clear", "close", "filters"]), n = typeof g == "function" ? g(t) : p(L, null, {
    default: () => [Array.isArray(F) && F.map((o, O) => {
      const v = (i) => {
        const f = i;
        f && (f.stopImmediatePropagation(), f.stopPropagation(), f.preventDefault());
        let s = [...y];
        const _ = s.includes(o.value), j = s.indexOf(o.value);
        return j > -1 ? s.splice(j, 1) : w ? s.push(o.value) : s = [o.value], c({
          value: o.value,
          filteredValue: s,
          included: !_,
          domEvent: f
        });
      }, P = y.includes(o.value), {
        text: d
      } = o, {
        value: m
      } = o, I = `${u}_${O}`, V = typeof C == "function" ? C({
        onChange: v,
        filterMultiple: w,
        value: m,
        text: d,
        checked: P,
        filteredValue: y,
        level: u
      }) : null;
      let D = V && k(V) ? q(V, {
        key: I
      }) : p(Q, {
        key: I,
        onClick: v
      }, {
        default: () => [w ? p(W, {
          checked: P
        }, A(d) ? d : {
          default: () => [d]
        }) : p(U, {
          checked: P
        }, A(d) ? d : {
          default: () => [d]
        })]
      });
      if (Array.isArray(o.children) && o.children.length) {
        const i = {
          ...e,
          filters: o.children,
          trigger: "hover",
          position: "right"
        };
        delete i.filterDropdownVisible, D = R(i, D, u + 1);
      }
      return D;
    })]
  }), r = {
    ...B(e, ...Object.keys(E)),
    onVisibleChange: (o) => l(o),
    trigger: x,
    position: S,
    render: n
  };
  return h != null && (r.visible = h), p(K, z(r, {
    key: `Dropdown_level_${u}`,
    className: `${M.PREFIX}-column-filter-dropdown`
  }), A(b) ? b : {
    default: () => [b]
  });
}
const ee = {
  ...E,
  prefixCls: String,
  filteredValue: Array,
  filterIcon: [Boolean, Object, Function],
  filterDropdown: [Boolean, Object, Function],
  filterDropdownProps: Object,
  filters: Array,
  filterMultiple: Boolean,
  filterDropdownVisible: Boolean,
  onSelect: Function,
  onFilterDropdownVisibleChange: Function,
  renderFilterDropdown: Function,
  renderFilterDropdownItem: Function,
  title: a.any,
  dataIndex: [Number, String],
  width: [Number, String],
  fixed: [Boolean, String],
  //@ts-ignore
  sortOrder: a.oneOfType([a.bool, a.string]),
  //@ts-ignore
  sorter: a.oneOfType([a.bool, a.func]),
  //@ts-ignore
  defaultFilteredValue: a.array
}, oe = {}, te = Y(ee, oe), pe = /* @__PURE__ */ X({
  props: te,
  name: "ColumnFilter",
  setup(e, {
    attrs: b
  }) {
    const {
      getProps: u
    } = Z(), w = typeof e.filterDropdownVisible < "u", F = typeof e.renderFilterDropdown == "function", h = !w && F ? !1 : e.filterDropdownVisible, c = T(e.filteredValue), l = T(h);
    N(() => e.filterDropdownVisible, () => {
      typeof e.filterDropdownVisible < "u" && (l.value = e.filterDropdownVisible);
    }, {
      immediate: !0
    }), N(() => e.filteredValue, () => {
      c.value = e.filteredValue;
    }, {
      immediate: !0
    });
    const x = (t = {}) => {
      const n = (t == null ? void 0 : t.filteredValue) || c;
      G(n, e.filteredValue) || e.onSelect({
        filteredValue: n
      }), t.closeDropdown && (l.value = !1);
    }, S = (t = {}) => {
      c.value = [], e.onSelect({
        filteredValue: []
      }), t.closeDropdown && (l.value = !1);
    }, g = () => {
      l.value = !1;
    }, C = (t) => {
      const n = typeof e.filterDropdownVisible < "u", r = typeof e.renderFilterDropdown == "function";
      !n && r && (l.value = t), e.onFilterDropdownVisibleChange(t);
    };
    return () => {
      const {
        prefixCls: t = M.PREFIX,
        filteredValue: n,
        filterIcon: r = "filter",
        filterDropdownProps: o,
        onSelect: O,
        filterDropdownVisible: v,
        renderFilterDropdown: P,
        onFilterDropdownVisibleChange: d
      } = u(e);
      let {
        filterDropdown: m = null
      } = e;
      const I = typeof e.filterDropdownVisible < "u", V = {
        tempFilteredValue: c.value,
        setTempFilteredValue: (s) => {
          c.value = s;
        },
        confirm: x,
        clear: S,
        close: g
      }, D = H(`${t}-column-filter`, {
        on: Array.isArray(n) && n.length
      });
      let i;
      typeof r == "function" ? i = r(Array.isArray(n) && n.length > 0) : k(r) ? i = r : i = p("div", {
        class: D
      }, ["​", p(J, {
        role: "button",
        "aria-label": "Filter data with this column",
        "aria-haspopup": "listbox",
        tabIndex: -1,
        size: "default"
      }, null)]);
      const f = {
        ...u(e),
        ...o,
        ...V,
        filterDropdownVisible: I ? v : l.value,
        onFilterDropdownVisibleChange: C
      };
      return m = k(m) ? m : R(f, i), m;
    };
  }
});
export {
  pe as ColumnFilter,
  te as ColumnFilterVueProps,
  pe as default
};
