// @kousum/semi-ui-vue/dist/modal/useModal/index.js
import m from "./HookModal";
import { withInfo as s, withSuccess as d, withError as h, withWarning as p, withConfirm as w } from "../confirm";
import { ref as g, nextTick as k, h as v } from "vue";
let f = 0;
function y() {
  const n = g([]);
  function t(e) {
    return n.value = [...Array.from(n.value), e], () => {
      n.value = Array.from(n.value).filter((o) => o !== e);
    };
  }
  return [n, t];
}
function x() {
  const [n, t] = y();
  function e(o) {
    return function(a) {
      f += 1;
      function c(u) {
        let i;
        const l = v(m, {
          key: `semi-modal-${f}`,
          config: o(u),
          afterClose: () => {
            i();
          }
        });
        return i = t(l), {
          modal: l,
          closeFunc: i
        };
      }
      let r = c(a);
      return {
        destroy: () => {
          r.closeFunc(), r = null;
        },
        update: (u) => {
          r && (r.closeFunc(), k(() => {
            r = c(u);
          }));
        }
      };
    };
  }
  return [{
    info: e(s),
    success: e(d),
    error: e(h),
    warning: e(p),
    confirm: e(w)
  }, n];
}
export {
  x as default
};
