// @kousum/semi-ui-vue/dist/configProvider/index.js
import { defineComponent as d, createVNode as n } from "vue";
import { BASE_CLASS_PREFIX as f } from '@douyinfe/semi-foundation/lib/es/base/constants';
import l from "../locale/source/zh_CN";
import o from "./context";
const s = o.Consumer, a = {
  locale: {
    type: Object,
    default: l
  },
  timeZone: String,
  getPopupContainer: Function,
  direction: {
    type: String,
    default: "ltr"
  }
}, g = /* @__PURE__ */ d({
  props: {
    ...a
  },
  name: "ConfigProvider",
  setup(r, {
    slots: e
  }) {
    function i() {
      const {
        direction: t
      } = r;
      return t === "rtl" ? n("div", {
        class: `${f}-rtl`
      }, [e.default ? e.default() : null]) : e.default ? e.default() : null;
    }
    return () => {
      const {
        direction: t,
        ...u
      } = r;
      return n(o.Provider, {
        value: {
          direction: t,
          ...u
        }
      }, {
        default: i
      });
    };
  }
});
export {
  s as ConfigConsumer,
  g as default,
  a as vuePropsType
};
