// @kousum/semi-ui-vue/dist/grid/col.js
import { defineComponent as i, inject as N, createVNode as g, mergeProps as h } from "vue";
import j from "classnames";
import { cssClasses as O } from '@douyinfe/semi-foundation/lib/es/grid/constants';
const y = {
  span: Number,
  order: Number,
  offset: Number,
  push: Number,
  pull: Number,
  className: String,
  prefixCls: {
    type: String,
    default: O.PREFIX
  },
  style: [Object, String],
  xs: [Number, Object],
  sm: [Number, Object],
  md: [Number, Object],
  lg: [Number, Object],
  xl: [Number, Object],
  xxl: [Number, Object]
}, v = /* @__PURE__ */ i({
  props: {
    ...y
  },
  name: "Col",
  setup(l, {
    slots: f
  }) {
    return () => {
      const {
        prefixCls: c,
        span: n,
        order: d,
        offset: m,
        push: a,
        pull: $,
        className: b,
        ...o
      } = l;
      let u = {};
      const t = `${c}-col`;
      ["xs", "sm", "md", "lg", "xl", "xxl"].forEach((s) => {
        let e = {};
        typeof l[s] == "number" ? e.span = l[s] : typeof l[s] == "object" && (e = l[s] || {}), delete o[s], u = {
          ...u,
          [`${t}-${s}-${e.span}`]: e.span !== void 0,
          [`${t}-${s}-order-${e.order}`]: e.order || e.order === 0,
          [`${t}-${s}-offset-${e.offset}`]: e.offset || e.offset === 0,
          [`${t}-${s}-push-${e.push}`]: e.push || e.push === 0,
          [`${t}-${s}-pull-${e.pull}`]: e.pull || e.pull === 0
        };
      });
      const x = j(t, {
        [`${t}-${n}`]: n !== void 0,
        [`${t}-order-${d}`]: d,
        [`${t}-offset-${m}`]: m,
        [`${t}-push-${a}`]: a,
        [`${t}-pull-${$}`]: $
      }, b, u);
      let {
        style: p
      } = o;
      const r = N("gutters", []);
      return p = {
        ...r[0] > 0 ? {
          paddingLeft: r[0] / 2 + "px",
          paddingRight: r[0] / 2 + "px"
        } : {},
        ...r[1] > 0 ? {
          paddingTop: r[1] / 2 + "px",
          paddingBottom: r[1] / 2 + "px"
        } : {},
        ...p
      }, g("div", h(o, {
        style: p,
        class: x
      }), [f.default ? f.default() : null]);
    };
  }
});
export {
  v as default,
  y as vuePropsType
};
