// @kousum/semi-ui-vue/dist/list/index.js
import { defineComponent as M, useSlots as V, useAttrs as _, cloneVNode as A, createVNode as o, mergeProps as x, isVNode as O } from "vue";
import B from "classnames";
import * as t from "../PropTypes";
import { vuePropsMake as D } from "../PropTypes";
import { cssClasses as a } from '@douyinfe/semi-foundation/lib/es/list/constants';
import { noop as I } from "lodash";
import '@douyinfe/semi-foundation/lib/es/list/list.css';
import K from "../locale/localeConsumer";
import q from "./item";
import { default as gt } from "./item";
import { Row as G } from "../grid";
import H from "../spin";
import J from "./list-context";
import Q from '@douyinfe/semi-foundation/lib/es/utils/getDataAttr';
function U(i) {
  return typeof i == "function" || Object.prototype.toString.call(i) === "[object Object]" && !O(i);
}
const s = a.PREFIX;
function W() {
  const i = {
    style: t.object,
    className: t.string,
    bordered: t.bool,
    footer: t.node,
    header: t.node,
    layout: t.string,
    size: t.string,
    split: t.bool,
    emptyContent: t.node,
    dataSource: t.array,
    renderItem: t.func,
    grid: t.object,
    loading: t.bool,
    loadMore: t.node,
    onRightClick: t.func,
    onClick: t.func
  }, R = D(i, {
    bordered: !1,
    split: !0,
    loading: !1,
    layout: "vertical",
    size: "default",
    onRightClick: I,
    onClick: I
  });
  return /* @__PURE__ */ M({
    props: {
      ...R
    },
    name: "List",
    setup(f, {}) {
      const d = V(), E = _(), k = () => {
        const {
          emptyContent: e
        } = f;
        return e ? o("div", {
          class: `${a.PREFIX}-empty`,
          "x-semi-prop": "emptyContent"
        }, [e]) : o(K, {
          componentName: "List"
        }, {
          default: (r) => o("div", {
            class: `${a.PREFIX}-empty`
          }, [r.emptyText])
        });
      };
      function L(e, r) {
        const {
          grid: l
        } = f;
        if (l) {
          const p = {};
          return ["align", "gutter", "justify", "type"].forEach((n) => {
            n in l && (p[n] = l[n]);
          }), o(G, x({
            type: "flex"
          }, p), {
            default: () => [e || null, r]
          });
        }
        return o("ul", {
          class: `${s}-items`
        }, [e || null, r]);
      }
      return () => {
        var v;
        let e;
        const r = (v = d.default) == null ? void 0 : v.call(d), {
          style: l,
          className: p,
          header: n,
          loading: g,
          onRightClick: j,
          onClick: N,
          footer: C,
          layout: S,
          grid: P,
          size: b,
          split: z,
          loadMore: $,
          bordered: F,
          dataSource: u,
          renderItem: h,
          ...X
        } = f, w = B(s, p, {
          [`${s}-flex`]: S === "horizontal",
          [`${s}-${b}`]: b,
          [`${s}-grid`]: P,
          [`${s}-split`]: z,
          [`${s}-bordered`]: F
        });
        let m;
        return u && u.length ? (m = [], (h ? u.map((c, y) => h(c, y)) : []).forEach((c, y) => {
          const T = c.key || `list-item-${y}`;
          m.push(A(c, {
            key: T
          }));
        })) : !r && !g && (m = k()), o("div", x({
          class: w,
          style: l
        }, Q({
          ...X,
          ...E
        })), [n ? o("div", {
          class: `${a.PREFIX}-header`,
          "x-semi-prop": "header"
        }, [n]) : null, o(J.Provider, {
          value: {
            grid: P,
            onRightClick: j,
            onClick: N
          }
        }, {
          default: () => [o(H, {
            spinning: g,
            size: "large"
          }, U(e = L(m, r)) ? e : {
            default: () => [e]
          })]
        }), C ? o("div", {
          class: `${a.PREFIX}-footer`,
          "x-semi-prop": "footer"
        }, [C]) : null, $ || null]);
      };
    }
  });
}
const Y = W(), Z = Y;
Z.Item = q;
export {
  gt as ListItem,
  W as List_,
  Z as default
};
