// @kousum/semi-ui-vue/dist/list/item.js
import { defineComponent as I, useSlots as S, useAttrs as T, createVNode as e, mergeProps as A, isVNode as G } from "vue";
import v from "classnames";
import * as t from "../PropTypes";
import { vuePropsMake as V } from "../PropTypes";
import { cssClasses as w } from '@douyinfe/semi-foundation/lib/es/list/constants';
import { noop as x } from "lodash";
import { Col as D } from "../grid";
import { useListContext as F } from "./context/Consumer";
import W from '@douyinfe/semi-foundation/lib/es/utils/getDataAttr';
function X(n) {
  return typeof n == "function" || Object.prototype.toString.call(n) === "[object Object]" && !G(n);
}
const o = w.PREFIX, _ = {
  extra: t.node,
  header: t.node,
  main: t.node,
  align: t.string,
  className: t.string,
  style: t.object,
  onClick: t.func,
  onRightClick: t.func,
  onMouseEnter: t.func,
  onMouseLeave: t.func
}, q = {
  align: "flex-start",
  onMouseEnter: x,
  onMouseLeave: x
}, z = V(_, q), tt = /* @__PURE__ */ I({
  props: {
    ...z
  },
  name: "ListItem",
  setup(n, {}) {
    const c = S(), h = T(), {
      context: a
    } = F();
    function k(s) {
      const {
        grid: i
      } = a.value, {
        gutter: r,
        justify: d,
        type: f,
        align: u,
        ...l
      } = i;
      return e(D, l, X(s) ? s : {
        default: () => [s]
      });
    }
    return () => {
      var y;
      const s = (y = c.default) == null ? void 0 : y.call(c), {
        header: i,
        main: r,
        className: d,
        style: f,
        extra: u,
        align: l,
        onClick: C,
        onRightClick: g,
        onMouseEnter: b,
        onMouseLeave: M,
        ...$
      } = n, {
        onRightClick: L,
        onClick: P,
        grid: j
      } = a.value, R = g || L, E = C || P, N = v(`${o}-item`, d), O = v(`${o}-item-body`, {
        [`${o}-item-body-${l}`]: l
      });
      let m;
      (i || r) && (m = e("div", {
        class: O
      }, [i ? e("div", {
        class: `${o}-item-body-header`
      }, [i]) : null, r ? e("div", {
        class: `${o}-item-body-main`
      }, [r]) : null]));
      let p = (
        // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-noninteractive-element-interactions
        e("li", A({
          class: N,
          style: f,
          onClick: E,
          onContextmenu: R,
          onMouseenter: b,
          onMouseleave: M
        }, W({
          ...$,
          ...h
        })), [m || null, s, u ? e("div", {
          class: `${o}-item-extra`
        }, [u]) : null])
      );
      return a.value && j && (p = k(p)), p;
    };
  }
});
export {
  tt as default,
  z as vuePropsType
};
