// @kousum/semi-ui-vue/dist/overflowList/__stories__/Demo.js
import { createVNode as e } from "vue";
import o from "../__test__/OverflowListDemo";
import r from "../__test__/OverflowListDemoScroll";
const n = {
  /* 👇 The title prop is optional.
   * See https://storybook.js.org/docs/7.0/vue/configure/overview#configure-story-loading
   * to learn how to generate automatic titles
   */
  title: "D展示类/OverflowList",
  render: (l) => ({
    setup() {
      return () => e("div", {
        style: {
          padding: "10px"
        }
      }, [l.name === "OverflowListDemo" ? e(o, null, null) : null, l.name === "OverflowListDemoScroll" ? e(r, null, null) : null]);
    }
  }),
  parameters: {
    // More on how to position stories at: https://storybook.js.org/docs/7.0/react/configure/story-layout
    layout: "fullscreen"
  },
  // This component will have an automatically generated docsPage entry: https://storybook.js.org/docs/7.0/vue/writing-docs/docs-page
  tags: ["autodocs"]
}, a = {
  args: {
    name: "OverflowListDemo"
  }
}, i = {
  args: {
    name: "OverflowListDemoScroll"
  }
};
export {
  a as OverflowList,
  i as OverflowListScroll,
  n as default
};
