// @kousum/semi-ui-vue/dist/steps/fillSteps.js
import { defineComponent as j, useSlots as $, createVNode as u, isVNode as C, cloneVNode as v } from "vue";
import * as t from "../PropTypes";
import { vuePropsMake as w } from "../PropTypes";
import F from "classnames";
import { stepsClasses as x } from '@douyinfe/semi-foundation/lib/es/steps/constants';
import { Row as T, Col as V } from "../grid";
import { getFragmentChildren as z } from "../_utils";
function y(e) {
  return typeof e == "function" || Object.prototype.toString.call(e) === "[object Object]" && !C(e);
}
const O = {
  prefixCls: t.string,
  className: t.string,
  style: t.object,
  current: t.number,
  initial: t.number,
  direction: t.string,
  status: t.string,
  children: t.node,
  onChange: t.func,
  size: t.string,
  "aria-label": t.string
}, R = {
  prefixCls: x.PREFIX,
  current: 0,
  direction: "horizontal",
  initial: 0,
  status: "process"
}, _ = w(O, R), D = /* @__PURE__ */ j({
  props: {
    ..._
  },
  name: "FillSteps",
  setup(e, {}) {
    const b = $();
    return () => {
      const {
        current: o,
        status: f,
        children: k,
        prefixCls: i,
        initial: m,
        direction: l,
        className: h,
        style: N,
        onChange: d
      } = e, a = () => {
        const g = z(b).filter((r) => C(r)), S = l === "vertical" ? null : {
          width: `${100 / g.length}%`
        };
        return g.map((r, n) => {
          let p;
          if (!r)
            return null;
          const c = m + n, s = {
            stepNumber: `${c + 1}`,
            direction: l,
            ...r.props
          };
          return f === "error" && n === o - 1 && (s.className = `${i}-next-error`), r.props.status || (c === o ? s.status = f : c < o ? s.status = "finish" : s.status = "wait"), s.onChange = d ? () => {
            n !== o && d(n + m);
          } : void 0, u(V, {
            style: S
          }, y(p = v(r, {
            ...s
          })) ? p : {
            default: () => [p]
          });
        });
      }, P = F(h, {
        [i]: !0,
        [`${i}-${l}`]: !0
      });
      return u("div", {
        class: P,
        style: N,
        "aria-label": e["aria-label"]
      }, [u(T, {
        type: "flex",
        justify: "start"
      }, y(a) ? a : {
        default: () => [a]
      })]);
    };
  }
});
export {
  D as default,
  _ as vuePropsType
};
