// @kousum/semi-ui-vue/dist/steps/index.js
import { defineComponent as u, useSlots as c, ref as f, createVNode as i, isVNode as d, mergeProps as l } from "vue";
import * as e from "../PropTypes";
import { vuePropsMake as m } from "../PropTypes";
import '@douyinfe/semi-foundation/lib/es/steps/steps.css';
import S from "./step";
import { default as K } from "./step";
import g from "./fillSteps";
import b from "./basicSteps";
import v from "./navSteps";
import y from "./context";
function h(t) {
  return typeof t == "function" || Object.prototype.toString.call(t) === "[object Object]" && !d(t);
}
const P = {
  onChange: e.func,
  // @ts-ignore
  type: e.string,
  size: e.string,
  current: e.number,
  style: e.object,
  direction: e.string,
  prefixCls: e.string,
  description: e.node,
  icon: e.node,
  status: e.string,
  title: e.node,
  className: e.string,
  onClick: e.func,
  active: e.bool,
  done: e.bool,
  children: e.node,
  stepNumber: [e.number, e.string],
  onKeyDown: e.func,
  role: e.string,
  "aria-label": e.string
}, C = {
  type: "fill",
  size: "default"
}, N = m(P, C), j = /* @__PURE__ */ u({
  props: {
    ...N
  },
  name: "Steps",
  setup(t, {}) {
    const r = c(), s = f([]);
    function a() {
      const {
        type: o,
        ...n
      } = t;
      switch (o) {
        case "fill":
          return i(g, l(n, {
            children: s.value
          }), {
            default: r.default
          });
        case "basic":
          return i(b, l(n, {
            children: s.value
          }), {
            default: r.default
          });
        case "nav":
          return i(v, l(n, {
            children: s.value
          }), {
            default: r.default
          });
        default:
          return null;
      }
    }
    return () => {
      var p;
      let o;
      s.value = (p = r.default) == null ? void 0 : p.call(r);
      const {
        type: n
      } = t;
      return i(y.Provider, {
        value: {
          type: n
        }
      }, h(o = a()) ? o : {
        default: () => [o]
      });
    };
  }
}), x = j;
x.Step = S;
export {
  K as StepsStep,
  x as default,
  N as vuePropsType
};
