// @kousum/semi-ui-vue/dist/button/__stories__/Button.js
import { createVNode as e, createTextVNode as r } from "vue";
import t from "../index";
const s = {
  /* 👇 The title prop is optional.
   * See https://storybook.js.org/docs/7.0/vue/configure/overview#configure-story-loading
   * to learn how to generate automatic titles
   */
  title: "B输入类/Button",
  component: t,
  render: (o) => ({
    components: {
      Button: t
    },
    setup() {
      return () => e("div", {
        style: {
          padding: "10px"
        }
      }, [e(t, o, {
        default: () => [r("123")]
      })]);
    }
  }),
  parameters: {
    // More on how to position stories at: https://storybook.js.org/docs/7.0/react/configure/story-layout
    layout: "fullscreen"
  },
  // This component will have an automatically generated docsPage entry: https://storybook.js.org/docs/7.0/vue/writing-docs/docs-page
  tags: ["autodocs"]
}, p = {
  args: {
    type: "primary"
  }
}, c = {
  args: {
    type: "secondary"
  }
};
export {
  p as Primary,
  c as Secondary,
  s as default
};
