// @kousum/semi-ui-vue/dist/tree/__stories__/Demo.js
import { createVNode as e } from "vue";
import o from "../__test__/TreeDemo";
import t from "../__test__/TreeCheckDemo";
const a = {
  /* 👇 The title prop is optional.
   * See https://storybook.js.org/docs/7.0/vue/configure/overview#configure-story-loading
   * to learn how to generate automatic titles
   */
  title: "C导航类/Tree",
  render: (r) => ({
    setup() {
      return () => e("div", {
        style: {
          padding: "10px"
        }
      }, [r.name === "TreeDemo" ? e(o, null, null) : null, r.name === "TreeCheckDemo" ? e(t, null, null) : null]);
    }
  }),
  parameters: {
    // More on how to position stories at: https://storybook.js.org/docs/7.0/react/configure/story-layout
    layout: "fullscreen"
  },
  // This component will have an automatically generated docsPage entry: https://storybook.js.org/docs/7.0/vue/writing-docs/docs-page
  tags: ["autodocs"]
}, u = {
  args: {
    name: "TreeDemo"
  }
}, s = {
  args: {
    name: "TreeCheckDemo"
  }
};
export {
  u as Tree,
  s as TreeCheck,
  a as default
};
