// @kousum/semi-illustrations-vue/dist/index.js
import { default as r } from "./illustrations/IllustrationConstructionDark";
import { default as l } from "./illustrations/IllustrationConstruction";
import { default as s } from "./illustrations/IllustrationFailureDark";
import { default as f } from "./illustrations/IllustrationFailure";
import { default as d } from "./illustrations/IllustrationIdleDark";
import { default as I } from "./illustrations/IllustrationIdle";
import { default as p } from "./illustrations/IllustrationNoAccessDark";
import { default as c } from "./illustrations/IllustrationNoAccess";
import { default as D } from "./illustrations/IllustrationNoContentDark";
import { default as C } from "./illustrations/IllustrationNoContent";
import { default as A } from "./illustrations/IllustrationNoResultDark";
import { default as S } from "./illustrations/IllustrationNoResult";
import { default as g } from "./illustrations/IllustrationNotFoundDark";
import { default as j } from "./illustrations/IllustrationNotFound";
import { default as v } from "./illustrations/IllustrationSuccessDark";
import { default as y } from "./illustrations/IllustrationSuccess";
export {
  l as IllustrationConstruction,
  r as IllustrationConstructionDark,
  f as IllustrationFailure,
  s as IllustrationFailureDark,
  I as IllustrationIdle,
  d as IllustrationIdleDark,
  c as IllustrationNoAccess,
  p as IllustrationNoAccessDark,
  C as IllustrationNoContent,
  D as IllustrationNoContentDark,
  S as IllustrationNoResult,
  A as IllustrationNoResultDark,
  j as IllustrationNotFound,
  g as IllustrationNotFoundDark,
  y as IllustrationSuccess,
  v as IllustrationSuccessDark
};
