// @kousum/semi-ui-vue/dist/timeline/index.js
import { defineComponent as I, useSlots as S, useAttrs as x, createVNode as u, mergeProps as d, isVNode as V, cloneVNode as k } from "vue";
import $ from "classnames";
import * as o from "../PropTypes";
import { vuePropsMake as v } from "../PropTypes";
import '@douyinfe/semi-foundation/lib/es/timeline/timeline.css';
import { cssClasses as A } from '@douyinfe/semi-foundation/lib/es/timeline/constants';
import g from "./item";
import { default as K } from "./item";
import j from '@douyinfe/semi-foundation/lib/es/utils/getDataAttr';
const s = A.PREFIX, B = {
  mode: o.string,
  className: o.string,
  style: o.object,
  dataSource: o.array,
  "aria-label": o.string
}, D = {
  mode: "left"
}, E = v(B, D), F = /* @__PURE__ */ I({
  props: {
    ...E
  },
  name: "Timeline",
  setup(i, {}) {
    const n = S(), y = x(), N = (t, e) => {
      const {
        mode: r
      } = i;
      return r === "alternate" ? t.props.position ? `${s}-item-${t.props.position}` : e % 2 === 0 ? `${s}-item-left` : `${s}-item-right` : r === "center" ? t.props.position ? `${s}-item-${t.props.position}` : `${s}-item-left` : r === "left" || r === "right" ? `${s}-item-${r}` : t.props.position ? `${s}-item-${t.props.position}` : "";
    }, m = (t) => t.map((e, r) => V(e) ? k(e, {
      // @ts-ignore
      className: $(e.props.class, N(e, r))
    }) : e);
    return () => {
      var l;
      const t = (l = n.default) == null ? void 0 : l.call(n), {
        className: e,
        style: r,
        mode: p,
        dataSource: a,
        ...P
      } = i, T = $(s, e, {
        [`${s}-${p}`]: p
      });
      let c;
      if (a && a.length) {
        const C = a.map((f, b) => u(g, d({
          key: `timeline-item-${b}`
        }, f), {
          default: () => [f.content]
        }));
        c = m(C);
      }
      const h = c || m(t);
      return u("ul", d({
        "aria-label": i["aria-label"],
        style: r,
        class: T
      }, j({
        ...P,
        ...y
      })), [h]);
    };
  }
}), L = F;
L.Item = g;
export {
  K as TimelineItem,
  L as default,
  E as vuePropsType
};
