// @kousum/semi-ui-vue/dist/timeline/item.js
import { defineComponent as T, useSlots as x, useAttrs as P, createVNode as o, mergeProps as m } from "vue";
import p from "classnames";
import { noop as I } from "lodash";
import * as t from "../PropTypes";
import { vuePropsMake as N } from "../PropTypes";
import { cssClasses as b } from '@douyinfe/semi-foundation/lib/es/timeline/constants';
import '@douyinfe/semi-foundation/lib/es/timeline/timeline.css';
import A from '@douyinfe/semi-foundation/lib/es/utils/getDataAttr';
const e = b.ITEM, M = {
  color: t.string,
  time: t.node,
  type: t.string,
  dot: t.node,
  extra: t.node,
  position: t.string,
  className: t.string,
  style: t.object,
  onClick: t.func
}, S = {
  type: "default",
  time: "",
  onClick: I
}, j = N(M, S), F = /* @__PURE__ */ T({
  props: {
    ...j
  },
  name: "TimelineItem",
  setup(d, {}) {
    const s = x(), u = P();
    return () => {
      var a;
      const f = (a = s.default) == null ? void 0 : a.call(s), {
        className: y,
        color: r,
        dot: n,
        type: i,
        style: C,
        time: l,
        extra: c,
        onClick: $,
        ...g
      } = d, v = p(e, y), h = p({
        [`${e}-head`]: !0,
        [`${e}-head-custom`]: n,
        [`${e}-head-${i}`]: i
      }), k = r ? {
        style: {
          backgroundColor: r
        }
      } : null;
      return o("li", m({
        class: v,
        style: C,
        onClick: $
      }, A({
        ...g,
        ...u
      })), [o("div", {
        class: `${e}-tail`,
        "aria-hidden": !0
      }, null), o("div", m({
        class: h,
        "aria-hidden": !0
      }, k), [n]), o("div", {
        class: `${e}-content`
      }, [f, c && o("div", {
        class: `${e}-content-extra`
      }, [c]), l && o("div", {
        class: `${e}-content-time`
      }, [l])])]);
    };
  }
});
export {
  F as default,
  j as vuePropsType
};
