// @kousum/semi-ui-vue/dist/transfer/__stories__/Demo.js
import { createVNode as e } from "vue";
import n from "../__test__/TransferDemo";
import a from "../__test__/TransferTreeDemo";
const l = {
  /* 👇 The title prop is optional.
   * See https://storybook.js.org/docs/7.0/vue/configure/overview#configure-story-loading
   * to learn how to generate automatic titles
   */
  title: "B输入类/Transfer",
  render: (r) => ({
    setup() {
      return () => e("div", {
        style: {
          padding: "10px"
        }
      }, [r.name === "TransferDemo" ? e(n, null, null) : null, r.name === "TransferTreeDemo" ? e(a, null, null) : null]);
    }
  }),
  parameters: {
    // More on how to position stories at: https://storybook.js.org/docs/7.0/react/configure/story-layout
    layout: "fullscreen"
  },
  // This component will have an automatically generated docsPage entry: https://storybook.js.org/docs/7.0/vue/writing-docs/docs-page
  tags: ["autodocs"]
}, m = {
  args: {
    name: "TransferDemo"
  }
}, f = {
  args: {
    name: "TransferTreeDemo"
  }
};
export {
  m as Transfer,
  f as TransferTree,
  l as default
};
