// @kousum/semi-ui-vue/dist/datePicker/footer.js
import { defineComponent as u, useSlots as m, createVNode as r, isVNode as C } from "vue";
import d from "classnames";
import l from "../button";
import { get as c } from "lodash";
import { vuePropsMake as k } from "../PropTypes";
function i(o) {
  return typeof o == "function" || Object.prototype.toString.call(o) === "[object Object]" && !C(o);
}
const y = {
  prefixCls: String,
  locale: {
    type: Object,
    required: !0
  },
  localeCode: {
    type: String,
    required: !0
  },
  onCancelClick: Function,
  onConfirmClick: Function
}, b = k(y, {}), O = /* @__PURE__ */ u({
  props: {
    ...b
  },
  name: "Footer",
  setup(o, {
    attrs: S
  }) {
    return m(), () => {
      let e, t;
      const {
        prefixCls: s,
        locale: n,
        onCancelClick: p,
        onConfirmClick: f
      } = o, a = d(`${s}-footer`);
      return r("div", {
        class: a
      }, [r(l, {
        theme: "borderless",
        onClick: p
      }, i(e = c(n, "footer.cancel", "")) ? e : {
        default: () => [e]
      }), r(l, {
        theme: "solid",
        onClick: f
      }, i(t = c(n, "footer.confirm", "")) ? t : {
        default: () => [t]
      })]);
    };
  }
});
export {
  O as default,
  b as vuePropsType
};
