// @kousum/semi-ui-vue/dist/datePicker/quickControl.js
import { defineComponent as x, useSlots as R, createVNode as e } from "vue";
import i from "classnames";
import { cssClasses as B } from '@douyinfe/semi-foundation/lib/es/datePicker/constants';
import w from "../button/index";
import { TypographyText as A } from "../typography/index";
import { noop as I } from '@douyinfe/semi-foundation/lib/es/utils/function';
import * as n from "../PropTypes";
import { vuePropsMake as M } from "../PropTypes";
const t = B.PREFIX, j = {
  presets: {
    type: n.array,
    required: !0
  },
  presetPosition: {
    type: n.string,
    required: !0
  },
  onPresetClick: {
    type: n.func,
    required: !0
  },
  type: {
    type: n.string,
    required: !0
  },
  insetInput: {
    type: [n.bool, n.object],
    required: !0
  },
  locale: {
    type: n.object,
    required: !0
  }
}, N = {
  presets: [],
  presetPosition: "bottom",
  onPresetClick: I
}, _ = M(j, N), X = /* @__PURE__ */ x({
  props: {
    ..._
  },
  name: "DatePicker_quickControl",
  setup(a, {}) {
    return R(), () => {
      const {
        presets: u,
        onPresetClick: m,
        type: p,
        presetPosition: o,
        insetInput: $,
        locale: d
      } = a, k = p === "dateRange" || p === "dateTimeRange", l = o === "top" || o === "bottom", q = p === "month", r = l && k, s = l && q, f = i(`${t}-quick-control`, {
        [`${t}-quick-control-${p}`]: p,
        [`${t}-quick-control-${o}`]: !0
      }), y = i({
        [`${t}-quick-control-header`]: !0
      }), C = i({
        [`${t}-quick-control-${o}-content-wrapper`]: !0
      }), P = i({
        [`${t}-quick-control-${o}-content`]: !r && !s,
        [`${t}-quick-control-${o}-range-content`]: r,
        [`${t}-quick-control-${o}-month-content`]: s
      }), g = i({
        [`${t}-quick-control-${o}-content-item`]: !r && !s,
        [`${t}-quick-control-${o}-range-content-item`]: r,
        [`${t}-quick-control-${o}-month-content-item`]: s
      }), h = i({
        [`${t}-quick-control-${o}-content-item-ellipsis`]: !r && !s,
        [`${t}-quick-control-${o}-range-content-item-ellipsis`]: r,
        [`${t}-quick-control-${o}-month-content-item-ellipsis`]: s
      });
      return u.length ? e("div", {
        class: f,
        "x-insetinput": $ ? "true" : "false"
      }, [!l && e("div", {
        class: y
      }, [d.presets]), e("div", {
        class: C
      }, [e("div", {
        class: P
      }, [u.map((c, T) => {
        const v = typeof c == "function" ? c() : c;
        return e(w, {
          size: "small",
          type: "primary",
          onClick: (b) => m(typeof c == "function" ? c() : c, b),
          key: T
        }, {
          default: () => [e("div", {
            class: g
          }, [e(A, {
            ellipsis: {
              showTooltip: !0
            },
            class: h
          }, {
            default: () => [v.text]
          })])]
        });
      })])])]) : null;
    };
  }
});
export {
  X as default,
  _ as vuePropsType
};
