// @kousum/semi-ui-vue/dist/navigation/CollapseButton.js
import { defineComponent as a, createVNode as n, isVNode as f } from "vue";
import { noop as u } from "lodash";
import { IconSidebar as d } from "@kousum/semi-icons-vue";
import p from "../button";
import C from "../tooltip";
function m(t) {
  return typeof t == "function" || Object.prototype.toString.call(t) === "[object Object]" && !f(t);
}
const x = {
  prefixCls: String,
  locale: Object,
  collapseText: Function,
  isCollapsed: {
    type: Boolean,
    default: void 0
  },
  onClick: Function
}, g = /* @__PURE__ */ a({
  props: {
    ...x
  },
  name: "CollapseButton",
  setup(t, b) {
    const r = () => {
      typeof t.onClick == "function" && t.onClick(!t.isCollapsed);
    }, i = {
      icon: n(d, null, null),
      type: "tertiary",
      theme: "borderless",
      onClick: r
    };
    return () => {
      const {
        prefixCls: s,
        locale: o,
        collapseText: c,
        isCollapsed: l,
        onClick: y = u
      } = t;
      let e = l ? o == null ? void 0 : o.expandText : o == null ? void 0 : o.collapseText;
      return typeof c == "function" && (e = c(l)), n("div", {
        class: `${s}-collapse-btn`
      }, [l ? n(C, {
        content: e,
        position: "right"
      }, {
        default: () => [n(p, i, null)]
      }) : n(p, i, m(e) ? e : {
        default: () => [e]
      })]);
    };
  }
});
export {
  g as default
};
