// @kousum/semi-ui-vue/dist/navigation/Footer.js
import { defineComponent as x, isVNode as f, createVNode as m, mergeProps as N } from "vue";
import * as t from "../PropTypes";
import v from "classnames";
import { strings as T, cssClasses as u } from '@douyinfe/semi-foundation/lib/es/navigation/constants';
import B from "./CollapseButton";
import '@douyinfe/semi-foundation/lib/es/navigation/navigation.css';
import { noop as P } from "lodash";
import { useNavContext as g } from "./nav-context/Consumer";
const k = {
  // children: PropTypes.node,
  style: t.object,
  className: t.string,
  collapseButton: {
    type: t.any,
    default: !1
  },
  collapseText: t.func,
  onClick: {
    type: t.func,
    default: P
  }
}, j = /* @__PURE__ */ x({
  props: {
    ...k
  },
  name: "NavFooter",
  setup(p, {
    slots: l
  }) {
    const {
      context: c
    } = g(), C = () => {
      const {
        collapseButton: e,
        collapseText: s
      } = p;
      if (f(e))
        return e;
      const {
        onCollapseChange: r,
        prefixCls: n,
        locale: o,
        isCollapsed: a
      } = c.value;
      return m(B, N({
        prefixCls: n,
        isCollapsed: a,
        locale: o,
        onClick: r,
        collapseText: s
      }, typeof e == "object" ? e : {}), null);
    };
    return () => {
      var i;
      const {
        style: e,
        className: s,
        collapseButton: r,
        onClick: n
      } = p;
      let o = (i = l.default) == null ? void 0 : i.call(l);
      const {
        isCollapsed: a,
        mode: d
      } = c.value;
      !f(o) && r && d !== T.MODE_HORIZONTAL && (o = C());
      const y = v(s, `${u.PREFIX}-footer`, {
        [`${u.PREFIX}-footer-collapsed`]: a
      });
      return m("div", {
        class: y,
        style: e,
        onClick: n
      }, [o]);
    };
  }
});
export {
  j as default,
  k as vuePropsType
};
