// @kousum/semi-ui-vue/dist/scrollList/index.js
import { defineComponent as b, useSlots as h, createVNode as e, mergeProps as y } from "vue";
import { useBaseComponent as S } from "../_base/baseComponent";
import { cssClasses as d } from '@douyinfe/semi-foundation/lib/es/scrollList/constants';
import m from "classnames";
import v from '@douyinfe/semi-foundation/lib/es/scrollList/foundation';
import '@douyinfe/semi-foundation/lib/es/scrollList/scrollList.css';
const j = {
  header: [Object, String],
  footer: [Object, String],
  children: [Object, String],
  bodyHeight: [Number, String],
  prefixCls: [String],
  className: String,
  style: Object
}, H = /* @__PURE__ */ b({
  props: {
    ...j
  },
  name: "ScrollList",
  setup(r, {}) {
    const s = h(), {
      adapter: p,
      getDataAttr: u
    } = S(r, {});
    return new v(p()), () => {
      var c;
      const {
        header: o,
        footer: n,
        prefixCls: a,
        bodyHeight: l,
        className: f,
        style: g,
        ...C
      } = r, t = m(f, {
        [a || d.PREFIX]: !0
      }), i = m({
        [`${a || d.PREFIX}-header`]: !0
      });
      return e("div", y({
        class: t,
        style: g
      }, u()), [o ? e("div", {
        class: i
      }, [e("div", {
        class: `${i}-title`
      }, [o]), e("div", {
        class: `${t}-line`
      }, null)]) : null, e("div", {
        class: `${t}-body`,
        style: {
          height: l || ""
        }
      }, [(c = s.default) == null ? void 0 : c.call(s)]), n ? e("div", {
        class: `${t}-footer`
      }, [n]) : null]);
    };
  }
});
export {
  H as default,
  j as vuePropsType
};
