// @kousum/semi-ui-vue/dist/avatar/avatarGroup.js
import { defineComponent as S, cloneVNode as $, createVNode as i, Fragment as h } from "vue";
import l from "classnames";
import { isNumber as M, get as E, isFunction as P } from "lodash";
import b from "./index";
import { strings as u, cssClasses as k } from '@douyinfe/semi-foundation/lib/es/avatar/constants';
import { getFragmentChildren as x } from "../_utils";
u.SIZE;
u.SHAPE;
u.OVERLAP_FROM;
const o = k.PREFIX, O = {
  style: [Object, String],
  className: String,
  shape: {
    type: String,
    default: "circle"
  },
  size: {
    type: String,
    default: "medium"
  },
  overlapFrom: {
    type: String,
    default: "start"
  },
  maxCount: Number,
  renderMore: Function
}, j = /* @__PURE__ */ S({
  props: {
    ...O
  },
  name: "AvatarGroup",
  setup(a, {
    slots: A
  }) {
    function v(e) {
      const {
        maxCount: r
      } = a;
      let t = e;
      const n = e.length - r, s = e.slice(0, r), m = e.slice(r);
      if (n > 0) {
        const p = y(n, m);
        s.push(p), t = s;
      }
      return t;
    }
    function y(e, r) {
      const {
        renderMore: t
      } = a, n = l(`${o}-item-more`);
      let s = i(b, {
        className: n,
        key: "_+n"
      }, {
        default: () => [`+${e}`]
      });
      return P(t) && (s = i(h, {
        key: "_+n"
      }, [t(e, r)])), s;
    }
    return () => {
      const e = x(A);
      function r() {
        return Array.isArray(e) ? e : [e];
      }
      const {
        maxCount: t,
        overlapFrom: n,
        size: s,
        shape: m,
        renderMore: p,
        ...C
      } = a;
      let f;
      const F = l({
        [`${o}-group`]: !0
      });
      if (e) {
        const d = r();
        f = (M(t) ? v(d) : d).map((g, c) => {
          const N = l(E(g.props, "className"), {
            [`${o}-item-start-${c}`]: n === "start",
            [`${o}-item-end-${c}`]: n === "end"
          });
          return $(g, {
            ...C,
            class: N,
            size: s,
            shape: m,
            key: c
          });
        });
      }
      return i("div", {
        class: F
      }, [f]);
    };
  }
});
export {
  j as default,
  O as vuePropsType
};
