// @kousum/semi-ui-vue/dist/image/image.js
import { defineComponent as J, useSlots as K, reactive as Q, ref as W, watch as Y, createVNode as s, mergeProps as b } from "vue";
import * as t from "../PropTypes";
import { vuePropsMake as ee } from "../PropTypes";
import { cssClasses as te } from '@douyinfe/semi-foundation/lib/es/image/constants';
import y from "classnames";
import { IconUploadError as re } from "@kousum/semi-icons-vue";
import oe from "./previewInner";
import se from '@douyinfe/semi-foundation/lib/es/image/imageFoundation';
import { LocaleConsumerFunc as ne } from "../locale/localeConsumer";
import { omit as ie, isUndefined as ae, isObject as le, isBoolean as ce } from "lodash";
import { SkeletonImage as me } from "../skeleton";
import '@douyinfe/semi-foundation/lib/es/image/image.css';
import { usePreviewContext as de } from "./previewContext/Consumer";
import { useBaseComponent as ue } from "../_base/baseComponent";
import { styleNum as C } from "../_utils";
ne();
const a = te.PREFIX, fe = {
  style: t.object,
  className: t.string,
  src: t.string,
  width: t.oneOfType([t.string, t.number]),
  height: t.oneOfType([t.string, t.number]),
  alt: t.string,
  placeholder: t.node,
  fallback: t.node,
  preview: [t.bool, t.object],
  onLoad: t.func,
  onError: t.func,
  onClick: t.func,
  crossOrigin: t.string,
  imageID: t.number,
  maxZoom: t.number,
  minZoom: t.number,
  setDownloadName: t.func,
  imgStyle: t.object,
  imgCls: t.string
}, ge = {
  preview: !0
}, pe = ee(fe, ge), he = /* @__PURE__ */ J({
  props: {
    ...pe
  },
  name: "Image",
  setup(n, {}) {
    K();
    const {
      context: c
    } = de(), l = Q({
      src: "",
      loadStatus: "loading",
      previewVisible: !1
    }), {
      adapter: S
    } = ue(n, l);
    function I() {
      return {
        ...S(),
        getIsInGroup: () => u()
        // getContexts: () => context.value,
        // getContext: key => { // eslint-disable-line
        //   if (context.value && key) {
        //     // @ts-ignore
        //     return context.value[key];
        //   }
        // }
      };
    }
    const N = I(), m = new se(N), P = W();
    function k(e, r) {
      const i = {};
      if (e.src !== r.src && (i.src = e.src, i.loadStatus = "loading"), le(e.preview)) {
        const {
          visible: f
        } = e.preview;
        ce(f) && (i.previewVisible = f);
      }
      return i;
    }
    Y([() => n.src, () => n.preview], () => {
      const e = k({
        ...n
      }, {
        ...l
      });
      e && Object.keys(e).forEach((r) => {
        l[r] = e[r];
      });
    }, {
      immediate: !0,
      deep: !0
    });
    function u() {
      return !!(c && c.value.isGroup);
    }
    function x() {
      return c ? c.value.lazyLoad : !1;
    }
    const E = (e) => {
      m.handleClick(e);
    }, L = (e) => {
      m.handleLoaded(e);
    }, O = (e) => {
      m.handleError(e);
    }, D = (e) => {
      m.handlePreviewVisibleChange(e);
    }, j = () => {
      const {
        width: e,
        height: r
      } = n;
      return s(me, {
        style: {
          width: e,
          height: r
        }
      }, null);
    }, V = () => {
      const e = `${a}-status`;
      return s("div", {
        class: e
      }, [s(re, {
        size: "extra-large"
      }, null)]);
    }, $ = () => {
      const e = `${a}-status`, {
        placeholder: r
      } = n;
      return r ? s("div", {
        class: e
      }, [r]) : j();
    }, T = () => {
      const {
        fallback: e
      } = n, r = `${a}-status`, i = typeof e == "string" ? s("img", {
        style: {
          width: "100%",
          height: "100%"
        },
        src: e,
        alt: "fallback"
      }, null) : e;
      return e ? s("div", {
        class: r
      }, [i]) : V();
    }, B = () => {
      const {
        loadStatus: e
      } = l;
      return s("div", {
        class: `${a}-overlay`
      }, [e === "error" && T(), e === "loading" && $()]);
    };
    return () => {
      const {
        src: e,
        loadStatus: r,
        previewVisible: i
      } = l, {
        src: f,
        width: F,
        height: z,
        alt: G,
        style: U,
        className: _,
        crossOrigin: g,
        preview: o,
        fallback: we,
        placeholder: ve,
        imageID: be,
        setDownloadName: R,
        imgCls: p,
        imgStyle: Z,
        ...M
      } = n, {
        width: h,
        height: w
      } = {
        width: C(F),
        height: C(z)
      }, X = Object.assign({
        width: h,
        height: w
      }, U), q = y(a, _), v = r === "success" && o && !u(), A = o && r === "success", H = typeof o != "boolean" ? o.src ?? e : e, d = typeof o != "boolean" && v ? {
        ...ie(o, ["className", "style", "previewCls", "previewStyle"]),
        className: o == null ? void 0 : o.previewCls,
        style: o == null ? void 0 : o.previewStyle
      } : {};
      return (
        // eslint-disable jsx-a11y/no-static-element-interactions
        // eslint-disable jsx-a11y/click-events-have-key-events
        s("div", {
          style: X,
          class: q,
          onClick: E
        }, [s("img", b({
          ref: P
        }, M, {
          src: u() && x() ? void 0 : e,
          "data-src": e,
          alt: G,
          style: Z,
          class: y(`${a}-img`, {
            [`${a}-img-preview`]: A,
            [`${a}-img-error`]: r === "error",
            [p]: !!p
          }),
          width: h,
          height: w,
          crossorigin: g,
          onError: O,
          onLoad: L
        }), null), r !== "success" && B(), v && s(oe, b(d, {
          src: H,
          visible: i,
          onVisibleChange: D,
          crossOrigin: ae(g) ? d == null ? void 0 : d.crossOrigin : g,
          setDownloadName: R
        }), null)])
      );
    };
  }
});
he.isSemiImage = !0;
export {
  he as default,
  pe as vuePropsType
};
