// @kousum/semi-ui-vue/dist/image/previewFooter.js
import { defineComponent as q, useSlots as G, createVNode as s, createTextVNode as H, isVNode as J } from "vue";
import { IconChevronLeft as Q, IconChevronRight as U, IconMinus as Y, IconPlus as oo, IconRealSizeStroked as eo, IconWindowAdaptionStroked as no, IconRotate as to, IconDownload as io } from "@kousum/semi-icons-vue";
import * as e from "../PropTypes";
import { vuePropsMake as ro } from "../PropTypes";
import ao from "../tooltip";
import w from "../divider";
import so from "../slider";
import { cssClasses as k } from '@douyinfe/semi-foundation/lib/es/image/constants';
import f from "classnames";
import lo from '@douyinfe/semi-foundation/lib/es/image/previewFooterFoundation';
import { LocaleConsumerFunc as co } from "../locale/localeConsumer";
import { throttle as uo } from "lodash";
import { useBaseComponent as mo } from "../_base/baseComponent";
function po(r) {
  return typeof r == "function" || Object.prototype.toString.call(r) === "[object Object]" && !J(r);
}
k.PREFIX;
const c = `${k.PREFIX}-preview-footer`, fo = co(), go = {
  curPage: e.number,
  totalNum: e.number,
  disabledPrev: e.bool,
  disabledNext: e.bool,
  disableDownload: e.bool,
  className: e.string,
  zoom: e.number,
  ratio: e.string,
  prevTip: e.string,
  nextTip: e.string,
  zoomInTip: e.string,
  zoomOutTip: e.string,
  rotateTip: e.string,
  downloadTip: e.string,
  adaptiveTip: e.string,
  originTip: e.string,
  showTooltip: e.bool,
  onZoomIn: e.func,
  onZoomOut: e.func,
  onPrev: e.func,
  onNext: e.func,
  onAdjustRatio: e.func,
  // onRotateLeft: PropTypes.func,
  onDownload: e.func,
  min: e.number,
  max: e.number,
  step: e.number,
  onRotate: e.func,
  renderPreviewMenu: e.func,
  forwardRef: e.object,
  zIndex: e.number
}, bo = {
  min: 10,
  max: 500,
  step: 10,
  showTooltip: !1,
  disableDownload: !1
}, ho = ro(go, bo), $o = /* @__PURE__ */ q({
  props: {
    ...ho
  },
  name: "Footer",
  setup(r, {}) {
    G();
    const {
      adapter: C
    } = mo(r, {});
    function R() {
      return {
        ...C()
      };
    }
    const x = R(), m = new lo(x), g = (o) => {
      m.changeSliderValue(o);
    }, b = () => {
      g("minus");
    }, h = () => {
      g("plus");
    }, v = () => {
      m.handleRotate("left");
    }, P = () => {
      m.handleRotate("right");
    }, z = uo((o) => {
      m.handleValueChange(o);
    }, 50), T = () => {
      m.handleRatioClick();
    }, N = () => {
      const {
        min: o,
        max: n,
        step: t,
        curPage: i,
        totalNum: a,
        ratio: l,
        zoom: p,
        disabledPrev: L,
        disabledNext: j,
        disableDownload: B,
        onNext: E,
        onPrev: A,
        onDownload: X,
        renderPreviewMenu: K
      } = r, W = {
        min: o,
        max: n,
        step: t,
        curPage: i,
        totalNum: a,
        ratio: l,
        zoom: p,
        disabledPrev: L,
        disabledNext: j,
        disableDownload: B,
        onNext: E,
        onPrev: A,
        onDownload: X,
        onRotateLeft: v,
        onRotateRight: P,
        disabledZoomIn: p === n,
        disabledZoomOut: p === o,
        onRatioClick: T,
        onZoomIn: h,
        onZoomOut: b,
        menuItems: I()
      };
      return K(W);
    }, u = (o, n, t) => {
      const {
        showTooltip: i,
        zIndex: a
      } = r;
      return i ? s(ao, {
        content: n,
        key: `tooltip-${t}`,
        zIndex: a + 1
      }, po(o) ? o : {
        default: () => [o]
      }) : o;
    }, d = (o) => s(fo, {
      componentName: "Image"
    }, {
      default: (n) => n[o]
    }), y = () => {
      const {
        disabledPrev: o,
        onPrev: n,
        prevTip: t
      } = r, i = s(Q, {
        key: "chevron-left",
        role: "prev_bt",
        size: "large",
        className: o ? `${c}-disabled` : "",
        onClick: o ? void 0 : n
      }, null), a = t ?? d("prevTip");
      return u(i, a, "chevron-left");
    }, $ = () => {
      const {
        disabledNext: o,
        onNext: n,
        nextTip: t
      } = r, i = s(U, {
        key: "chevron-right",
        role: "next_bt",
        size: "large",
        className: o ? `${c}-disabled` : "",
        onClick: o ? void 0 : n
      }, null), a = t ?? d("nextTip");
      return u(i, a, "chevron-right");
    }, S = () => {
      const {
        zoomOutTip: o,
        zoom: n,
        min: t
      } = r, i = n === t, a = s(Y, {
        key: "minus",
        role: "minus_bt",
        size: "large",
        onClick: i ? void 0 : b,
        className: i ? `${c}-disabled` : ""
      }, null), l = o ?? d("zoomOutTip");
      return u(a, l, "minus");
    }, D = () => {
      const {
        zoomInTip: o,
        zoom: n,
        max: t
      } = r, i = n === t, a = s(oo, {
        key: "plus",
        role: "plus_bt",
        size: "large",
        onClick: i ? void 0 : h,
        className: i ? `${c}-disabled` : ""
      }, null), l = o ?? d("zoomInTip");
      return u(a, l, "plus");
    }, F = () => {
      const {
        ratio: o,
        originTip: n,
        adaptiveTip: t
      } = r, i = {
        key: "ratio",
        size: "large",
        className: f(`${c}-gap`),
        onClick: T
      }, a = o === "adaptation" ? s(eo, i, null) : s(no, i, null);
      let l;
      return o === "adaptation" ? l = n ?? d("originTip") : l = t ?? d("adaptiveTip"), u(a, l, "ratio");
    }, M = () => {
      const {
        rotateTip: o
      } = r, n = s(to, {
        key: "rotate",
        role: "rotate_bt",
        size: "large",
        onClick: v
      }, null), t = o ?? d("rotateTip");
      return u(n, t, "rotate");
    }, _ = () => {
      const {
        downloadTip: o,
        onDownload: n,
        disableDownload: t
      } = r, i = s(io, {
        key: "download",
        role: "download_bt",
        size: "large",
        onClick: t ? void 0 : n,
        className: f(`${c}-gap`, {
          [`${c}-disabled`]: t
        })
      }, null), a = o ?? d("downloadTip");
      return u(i, a, "download");
    }, O = () => {
      const {
        curPage: o,
        totalNum: n
      } = r;
      return s("div", {
        class: `${c}-page`,
        key: "info"
      }, [o, H("/"), n]);
    }, V = () => {
      const {
        zoom: o,
        min: n,
        max: t,
        step: i,
        showTooltip: a
      } = r;
      return s(so, {
        key: "slider",
        value: o,
        min: n,
        max: t,
        step: i,
        tipFormatter: (l) => `${l}%`,
        tooltipVisible: a ? void 0 : !1,
        onChange: z
      }, null);
    }, I = () => [y(), O(), $(), S(), V(), D(), F(), M(), _()], Z = () => {
      const o = I();
      return o.splice(3, 0, s(w, {
        layout: "vertical",
        key: "divider-first"
      }, null)), o.splice(8, 0, s(w, {
        layout: "vertical",
        key: "divider-second"
      }, null)), o;
    };
    return () => {
      const {
        className: o,
        renderPreviewMenu: n,
        forwardRef: t
      } = r, i = f(c, `${c}-wrapper`, o, {
        [`${c}-content`]: !n
      });
      return s("section", {
        class: i,
        ref: t
      }, [n ? N() : Z()]);
    };
  }
});
export {
  $o as default,
  ho as vuePropsType
};
