// @kousum/semi-ui-vue/dist/tagInput/SortableItem.js
import { defineComponent as s, shallowRef as l, watch as m, ref as a } from "vue";
import { useSortable as c } from "@kousum/dnd-kit-vue/sortable";
import { pointerIntersection as f } from "@dnd-kit/collision";
const b = {
  id: {
    type: [String, Number],
    required: !0
  },
  item: {
    type: Function,
    required: !0
  },
  index: {
    type: Number,
    required: !0
  }
}, y = /* @__PURE__ */ s({
  props: {
    ...b
  },
  name: "SortableItem",
  setup(t, {}) {
    const r = l(t.id);
    m(() => t.id, (e, n) => {
      e !== n && (r.value = n);
    });
    const o = l(t.index);
    m(() => t.index, (e, n) => {
      e !== n && (o.value = n);
    });
    const u = a(null), d = a(null), {
      isDragSource: i
    } = c({
      id: r,
      index: o,
      element: u,
      handle: d,
      collisionDetector: f
    });
    return () => t.item({
      element: u,
      handleRef: (e) => {
        d.value = e == null ? void 0 : e.$el;
      },
      attributes: {
        shadow: i == null ? void 0 : i.value
      }
    });
  }
});
export {
  y as default,
  b as vuePropsType
};
