// @kousum/semi-ui-vue/dist/tagInput/SortableList.js
import { defineComponent as i, useSlots as l, ref as s, createVNode as n, isVNode as u } from "vue";
import d from "./SortableItem";
import { DragDropProvider as c } from "@kousum/dnd-kit-vue";
function p(o) {
  return typeof o == "function" || Object.prototype.toString.call(o) === "[object Object]" && !u(o);
}
const S = {
  items: Array,
  onSortOver: Function,
  onSortEnd: Function,
  useDragHandle: [Boolean],
  helperClass: String,
  axis: String
}, y = /* @__PURE__ */ i({
  props: {
    ...S
  },
  name: "SortableList",
  setup(o, {}) {
    l();
    const a = s("");
    return () => {
      let r;
      return n(c, {
        onDragStart: (e) => {
          var t;
          a.value = "" + ((t = e.operation.source) == null ? void 0 : t.id);
        },
        onDragOver: (e) => {
          o.onSortOver(e);
        },
        onDragEnd: (e) => {
          o.onSortEnd(e);
        }
      }, p(r = o.items.map((e, t) => n(d, {
        key: e.key,
        id: e.key,
        item: e.item,
        index: t
      }, null))) ? r : {
        default: () => [r]
      });
    };
  }
});
export {
  y as default,
  S as vuePropsType
};
