// @kousum/semi-ui-vue/dist/badge/index.js
import { defineComponent as k, useSlots as w, createVNode as m, mergeProps as S } from "vue";
import f from "classnames";
import * as o from "../PropTypes";
import { vuePropsMake as E } from "../PropTypes";
import { isNumber as d, isString as j, noop as i } from "lodash";
import { cssClasses as I } from '@douyinfe/semi-foundation/lib/es/badge/constants';
import '@douyinfe/semi-foundation/lib/es/badge/badge.css';
import { useConfigContext as L } from "../configProvider/context/Consumer";
import { useHasInProps as F } from "../_base/baseComponent";
const e = I.PREFIX, H = {
  count: o.node,
  dot: o.bool,
  type: o.string,
  theme: o.string,
  position: o.string,
  overflowCount: o.number,
  style: o.object,
  className: o.string,
  onClick: o.func,
  onMouseEnter: o.func,
  onMouseLeave: o.func,
  countClassName: o.string,
  countStyle: o.object
}, R = {
  dot: !1,
  type: "primary",
  theme: "solid",
  className: "",
  onClick: () => i,
  onMouseEnter: () => i,
  onMouseLeave: () => i
}, V = E(H, R), O = /* @__PURE__ */ k({
  props: {
    ...V
  },
  name: "Badge",
  setup(g, {}) {
    const {
      getProps: y
    } = F(), n = w(), {
      context: $
    } = L();
    return () => {
      var a;
      const {
        direction: C
      } = $.value, P = C === "rtl" ? "leftTop" : "rightTop", {
        count: t,
        dot: r,
        type: v,
        countClassName: N,
        countStyle: b,
        theme: h,
        position: u = P,
        overflowCount: c,
        className: x,
        ...B
      } = y(g), p = (a = n.default) == null ? void 0 : a.call(n), s = t && !(d(t) || j(t)), M = t !== null && typeof t < "u", T = f(N, {
        [`${e}-${v}`]: !s,
        [`${e}-${h}`]: !s,
        [`${e}-${u}`]: !!u && !!p,
        [`${e}-block`]: !p,
        [`${e}-dot`]: r,
        [`${e}-count`]: !r && !s && M,
        [`${e}-custom`]: s
      });
      let l;
      return d(t) ? l = c && c < t ? `${c}+` : `${t}` : l = t, m("span", S({
        class: f(e, x)
      }, B), [p, m("span", {
        class: T,
        style: b,
        "x-semi-prop": "count"
      }, [r ? null : l])]);
    };
  }
});
export {
  O as default,
  V as vuePropsType
};
