// @kousum/semi-ui-vue/dist/carousel/CarouselIndicator.js
import { defineComponent as f, useSlots as O, createVNode as C, mergeProps as R } from "vue";
import l from "classnames";
import * as t from "../PropTypes";
import { vuePropsMake as N } from "../PropTypes";
import { cssClasses as r } from '@douyinfe/semi-foundation/lib/es/carousel/constants';
import h from '@douyinfe/semi-foundation/lib/es/utils/getDataAttr';
const v = {
  activeKey: t.number,
  className: t.string,
  position: t.string,
  size: t.string,
  style: t.object,
  theme: t.string,
  total: t.number,
  onIndicatorChange: t.func,
  type: String,
  trigger: t.string,
  activeIndex: t.number,
  defaultActiveIndex: t.number
}, y = N(v, {}), L = /* @__PURE__ */ f({
  props: {
    ...y
  },
  name: "CarouselIndicator",
  setup(s, {}) {
    O();
    const m = (e) => {
      s.onIndicatorChange(e);
    }, I = (e) => {
      const {
        trigger: o
      } = s;
      o === "click" && m(e);
    }, d = (e) => {
      const {
        trigger: o
      } = s;
      o === "hover" && m(e);
    };
    function u() {
      const {
        total: e,
        theme: o,
        size: a,
        activeIndex: c
      } = s, i = [];
      for (let n = 0; n < e; n++)
        i.push(C("span", {
          key: n,
          "data-index": n,
          class: l([`${r.CAROUSEL_INDICATOR}-item`], {
            [`${r.CAROUSEL_INDICATOR}-item-active`]: n === c,
            [`${r.CAROUSEL_INDICATOR}-item-${o}`]: o,
            [`${r.CAROUSEL_INDICATOR}-item-${a}`]: a
          }),
          onClick: () => I(n),
          onMouseenter: () => d(n)
        }, null));
      return i;
    }
    return () => {
      const {
        type: e,
        size: o,
        theme: a,
        style: c,
        className: i,
        position: n,
        ...p
      } = s, g = l(i, {
        [r.CAROUSEL_INDICATOR]: !0,
        [`${r.CAROUSEL_INDICATOR}-${e}`]: e,
        [`${r.CAROUSEL_INDICATOR}-${n}`]: n
      }), A = u();
      return C("div", R({
        class: g,
        style: c
      }, h(p)), [A]);
    };
  }
});
export {
  L as default,
  y as vuePropsType
};
