// @kousum/semi-ui-vue/dist/carousel/index.js
import { defineComponent as k, useSlots as B, shallowRef as W, reactive as X, watch as q, onMounted as G, onBeforeUnmount as H, createVNode as p, mergeProps as J, Fragment as K, cloneVNode as Q } from "vue";
import I from "classnames";
import * as t from "../PropTypes";
import { vuePropsMake as Y } from "../PropTypes";
import { useBaseComponent as Z } from "../_base/baseComponent";
import { numbers as h, cssClasses as s } from '@douyinfe/semi-foundation/lib/es/carousel/constants';
import ee from '@douyinfe/semi-foundation/lib/es/carousel/foundation';
import te from "./CarouselIndicator";
import ne from "./CarouselArrow";
import '@douyinfe/semi-foundation/lib/es/carousel/carousel.css';
import { debounce as y } from "lodash";
import oe from '@douyinfe/semi-foundation/lib/es/utils/isNullOrUndefined';
import { getFragmentChildren as re } from "../_utils";
const ie = {
  activeIndex: t.number,
  animation: t.string,
  arrowProps: t.object,
  autoPlay: t.oneOfType([t.bool, t.object]),
  className: t.string,
  defaultActiveIndex: t.number,
  indicatorPosition: t.string,
  indicatorSize: t.string,
  indicatorType: t.string,
  theme: t.string,
  onChange: t.func,
  arrowType: t.string,
  showArrow: t.bool,
  showIndicator: t.bool,
  slideDirection: t.string,
  speed: t.number,
  style: t.object,
  trigger: t.string
}, se = {
  // children: [],
  animation: "slide",
  autoPlay: !0,
  arrowType: "always",
  defaultActiveIndex: h.DEFAULT_ACTIVE_INDEX,
  indicatorPosition: "center",
  indicatorSize: "small",
  indicatorType: "dot",
  theme: "light",
  onChange: () => {
  },
  showArrow: !0,
  showIndicator: !0,
  slideDirection: "left",
  speed: h.DEFAULT_SPEED,
  trigger: "click"
}, ae = Y(ie, se), he = /* @__PURE__ */ k({
  props: {
    ...ae
  },
  name: "Carousel",
  setup(a, {
    expose: A
  }) {
    const x = B();
    W({
      children: []
    });
    let u = [];
    const r = X({
      activeIndex: -1,
      children: [],
      preIndex: -1,
      isReverse: !1,
      isInit: !0
    }), {
      adapter: E,
      getDataAttr: N
    } = Z(a, r);
    function P() {
      return {
        ...E(),
        getStates() {
          return {
            ...r,
            children: u
          };
        },
        notifyChange: (e, n) => {
          a.onChange(e, n);
        },
        setNewActiveIndex: (e) => {
          r.activeIndex = e;
        },
        setPreActiveIndex: (e) => {
          r.preIndex = e;
        },
        setIsReverse: (e) => {
          r.isReverse = e;
        },
        setIsInit: (e) => {
          r.isInit = e;
        }
      };
    }
    const O = P(), o = new ee(O), g = o.getDefaultActiveIndex();
    r.activeIndex = g, r.preIndex = g;
    function w(e) {
      const n = {};
      return !oe(e.activeIndex) && e.activeIndex !== r.activeIndex && (n.activeIndex = e.activeIndex), n;
    }
    q([() => a.activeIndex, () => r.activeIndex], (e) => {
      const n = w({
        ...a
      });
      n && Object.keys(n).forEach((i) => {
        r[i] = n[i];
      });
    }), G(() => {
      L();
    }), H(() => {
      o.destroy();
    });
    const S = () => (o.setForcePlay(!0), o.handleAutoPlay()), R = () => (o.setForcePlay(!1), o.stop()), U = (e) => o.goTo(e), f = () => o.prev(), v = () => o.next();
    A({
      play: S,
      stop: R,
      goTo: U,
      prev: f,
      next: v
    });
    const L = () => {
      o.getIsControlledComponent() || o.handleAutoPlay();
    }, _ = () => {
      const {
        autoPlay: e
      } = a;
      (e === !0 || typeof e == "object" && e.hoverToPause) && o.stop();
    }, b = () => {
      const {
        autoPlay: e
      } = a;
      (typeof e != "object" || e.hoverToPause) && !o.getIsControlledComponent() && o.handleAutoPlay();
    }, D = (e) => o.onIndicatorChange(e), T = (e) => o.getValidIndex(e), $ = () => {
      const {
        speed: e,
        animation: n
      } = a, {
        activeIndex: i,
        preIndex: c,
        isInit: l
      } = r;
      return p(K, null, [u.map((d, m) => {
        const C = m === i, M = m === T(i - 1), V = m === T(i + 1), z = {
          transitionTimingFunction: "ease",
          transitionDuration: `${e}ms`,
          animationTimingFunction: "ease",
          animationDuration: `${e}ms`
        };
        return Q(d, {
          style: {
            ...d.props.style,
            ...z
          },
          className: I(d.props.class, {
            [`${s.CAROUSEL_CONTENT}-item-prev`]: M,
            [`${s.CAROUSEL_CONTENT}-item-next`]: V,
            [`${s.CAROUSEL_CONTENT}-item-current`]: C,
            [`${s.CAROUSEL_CONTENT}-item`]: !0,
            [`${s.CAROUSEL_CONTENT}-item-active`]: C,
            [`${s.CAROUSEL_CONTENT}-item-slide-in`]: n === "slide" && !l && C,
            [`${s.CAROUSEL_CONTENT}-item-slide-out`]: n === "slide" && !l && m === c
          })
        });
      })]);
    }, F = () => {
      const {
        activeIndex: e
      } = r, {
        showIndicator: n,
        indicatorType: i,
        theme: c,
        indicatorPosition: l,
        indicatorSize: d,
        trigger: m
      } = a, C = I({
        [s.CAROUSEL_INDICATOR]: !0
      });
      return n && u.length > 1 ? p("div", {
        class: C
      }, [p(te, {
        type: i,
        total: u.length,
        activeIndex: e,
        position: l,
        trigger: m,
        size: d,
        theme: c,
        onIndicatorChange: D
      }, null)]) : null;
    }, j = () => {
      const {
        showArrow: e,
        arrowType: n,
        theme: i,
        arrowProps: c
      } = a;
      return e && u.length > 1 ? p(ne, {
        type: n,
        theme: i,
        prev: f,
        next: v,
        arrowProps: c
      }, null) : null;
    };
    return () => {
      u = re(x);
      const {
        animation: e,
        className: n,
        style: i,
        slideDirection: c
      } = a, {
        isReverse: l
      } = r, d = I(n, {
        [s.CAROUSEL]: !0
      });
      return p("div", J({
        class: d,
        style: i,
        onMouseenter: y(_, 400),
        onMouseleave: y(b, 400)
      }, N()), [p("div", {
        class: I([`${s.CAROUSEL_CONTENT}-${e}`], {
          [`${s.CAROUSEL_CONTENT}`]: !0,
          [`${s.CAROUSEL_CONTENT}-reverse`]: c === "left" ? l : !l
        }),
        "x-semi-prop": "children"
      }, [$()]), F(), j()]);
    };
  }
});
export {
  he as default,
  ae as vuePropsType
};
