// @kousum/semi-ui-vue/dist/rating/item.js
import { defineComponent as G, useSlots as J, reactive as Q, ref as z, createVNode as l, mergeProps as I } from "vue";
import b from "classnames";
import * as t from "../PropTypes";
import { vuePropsMake as U } from "../PropTypes";
import { cssClasses as S } from '@douyinfe/semi-foundation/lib/es/rating/constants';
import '@douyinfe/semi-foundation/lib/es/rating/rating.css';
import { IconStar as Y } from "@kousum/semi-icons-vue";
import { RatingItemFoundation as Z } from '@douyinfe/semi-foundation/lib/es/rating/foundation';
import { useBaseComponent as ee } from "../_base/baseComponent";
const re = {
  value: {
    type: t.number,
    required: !0
  },
  index: {
    type: t.number,
    required: !0
  },
  prefixCls: {
    type: t.string,
    required: !0
  },
  allowHalf: {
    type: t.bool,
    required: !0
  },
  onHover: {
    type: t.func,
    required: !0
  },
  onClick: {
    type: t.func,
    required: !0
  },
  character: {
    type: t.node,
    required: !0
  },
  focused: {
    type: t.bool,
    required: !0
  },
  disabled: {
    type: t.bool,
    required: !0
  },
  count: {
    type: t.number,
    required: !0
  },
  ariaLabelPrefix: {
    type: t.string,
    required: !0
  },
  size: {
    type: t.oneOfType([t.string, t.number]),
    required: !0
  },
  "aria-describedby": t.string,
  onFocus: t.func,
  onBlur: t.func,
  preventScroll: t.bool,
  style: t.object,
  className: t.string
}, te = U(re, {}), pe = /* @__PURE__ */ G({
  props: {
    ...te
  },
  name: "Item",
  setup(o, {
    expose: g
  }) {
    J();
    const u = Q({
      firstStarFocus: !1,
      secondStarFocus: !1
    }), {
      adapter: w
    } = ee(o, u);
    function B() {
      return {
        ...w(),
        setFirstStarFocus: (e) => {
          u.firstStarFocus = e;
        },
        setSecondStarFocus: (e) => {
          u.secondStarFocus = e;
        }
      };
    }
    const H = B(), F = new Z(H), v = z(), $ = z(), T = (e) => {
      const {
        onHover: r,
        index: s
      } = o;
      r(e, s);
    }, E = (e) => {
      const {
        onClick: r,
        index: s
      } = o;
      r(e, s);
    }, x = (e, r) => {
      const {
        onFocus: s
      } = o;
      s && s(e), F.handleFocusVisible(e, r);
    }, h = (e, r) => {
      const {
        onBlur: s
      } = o;
      s && s(e), F.handleBlur(e, r);
    }, R = (e) => {
      const {
        onClick: r,
        index: s
      } = o;
      e.keyCode === 13 && r(e, s);
    };
    return g({
      starFocus: () => {
        const {
          value: e,
          index: r,
          preventScroll: s
        } = o;
        e - r === 0.5 ? v.value.focus({
          preventScroll: s
        }) : $.value.focus({
          preventScroll: s
        });
      }
    }), () => {
      const {
        index: e,
        prefixCls: r,
        character: s,
        count: p,
        value: a,
        disabled: n,
        allowHalf: d,
        focused: oe,
        size: c,
        ariaLabelPrefix: C
      } = o, {
        firstStarFocus: N,
        secondStarFocus: V
      } = u, q = e + 1, y = q - a, W = d && y < 1 && y > 0, X = 1 - y, j = q <= a, m = typeof c == "number", K = b(r, {
        [`${r}-half`]: W,
        [`${r}-full`]: j,
        [`${r}-${c}`]: !m
      }), L = m ? {
        width: c,
        height: c,
        fontSize: c
      } : {}, P = s || l(Y, {
        size: m ? "inherit" : c === "small" ? "default" : "extra-large",
        style: {
          display: "block"
        }
      }, null), f = e === p, M = b(`${r}-wrapper`, {
        [`${r}-disabled`]: n,
        [`${S.PREFIX}-focus`]: (N || V) && a !== 0
      }), A = {
        onClick: n ? null : E,
        onKeydown: n ? null : R,
        onMousemove: n ? null : T,
        class: M
      }, k = d ? p * 2 + 1 : p + 1, D = {
        ref: v,
        role: "radio",
        "aria-checked": a === e + 0.5,
        "aria-posinset": 2 * e + 1,
        "aria-setsize": k,
        "aria-disabled": n,
        "aria-label": `${e + 0.5} ${C}s`,
        "aria-labelledby": o["aria-describedby"],
        "aria-describedby": o["aria-describedby"],
        class: b(`${r}-first`, `${S.PREFIX}-no-focus`),
        tabindex: !n && a === e + 0.5 ? 0 : -1,
        onFocus: (i) => {
          x(i, "first");
        },
        onBlur: (i) => {
          h(i, "first");
        }
      }, O = !n && (a === e + 1 || f && a === 0) ? 0 : -1, _ = {
        ref: $,
        role: "radio",
        "aria-checked": f ? a === 0 : a === e + 1,
        "aria-posinset": d ? 2 * (e + 1) : e + 1,
        "aria-setsize": k,
        "aria-disabled": n,
        "aria-label": `${f ? 0 : e + 1} ${C}${e === 0 ? "" : "s"}`,
        "aria-labelledby": o["aria-describedby"],
        "aria-describedby": o["aria-describedby"],
        className: b(`${r}-second`, `${S.PREFIX}-no-focus`),
        tabIndex: O,
        onFocus: (i) => {
          x(i, "second");
        },
        onBlur: (i) => {
          h(i, "second");
        }
      };
      return l("li", {
        class: K,
        style: {
          ...L
        },
        key: e
      }, [l("div", A, [d && !f && l("div", I(D, {
        style: {
          width: `${X * 100}%`
        }
      }), [P]), l("div", I(_, {
        "x-semi-prop": "character"
      }), [P])])]);
    };
  }
});
export {
  pe as default,
  te as vuePropsType
};
