// @kousum/semi-ui-vue/dist/table/__stories__/Demo.js
import { createVNode as l } from "vue";
import m from "../__test__/TableDemo1";
import a from "../__test__/TableDemo2";
import o from "../__test__/TableDemo3";
import n from "../__test__/TableDemo4";
import b from "../__test__/TableDemo5";
import T from "../__test__/TableDemo6";
import u from "../__test__/TableDemo7";
import t from "../__test__/TableDemo8";
import r from "../__test__/TableDemo9";
import D from "../__test__/TableDemo10";
import i from "../__test__/TableDemo11";
import s from "../__test__/TableDemo12";
import p from "../__test__/TableDemo13Virtualized";
const w = {
  /* 👇 The title prop is optional.
   * See https://storybook.js.org/docs/7.0/vue/configure/overview#configure-story-loading
   * to learn how to generate automatic titles
   */
  title: "D展示类/Table",
  render: (e) => ({
    setup() {
      return () => l("div", {
        style: {
          padding: "10px"
        }
      }, [e.name === "TableDemo1" ? l(m, null, null) : null, e.name === "TableDemo2" ? l(a, null, null) : null, e.name === "TableDemo3" ? l(o, null, null) : null, e.name === "TableDemo4" ? l(n, null, null) : null, e.name === "TableDemo5" ? l(b, null, null) : null, e.name === "TableDemo6" ? l(T, null, null) : null, e.name === "TableDemo7" ? l(u, null, null) : null, e.name === "TableDemo8" ? l(t, null, null) : null, e.name === "TableDemo9" ? l(r, null, null) : null, e.name === "TableDemo10" ? l(D, null, null) : null, e.name === "TableDemo11" ? l(i, null, null) : null, e.name === "TableDemo12" ? l(s, null, null) : null, e.name === "TableDemo13Virtualized" ? l(p, null, null) : null]);
    }
  }),
  parameters: {
    // More on how to position stories at: https://storybook.js.org/docs/7.0/react/configure/story-layout
    layout: "fullscreen"
  },
  // This component will have an automatically generated docsPage entry: https://storybook.js.org/docs/7.0/vue/writing-docs/docs-page
  tags: ["autodocs"]
}, A = {
  args: {
    name: "TableDemo1"
  }
}, B = {
  args: {
    name: "TableDemo2"
  }
}, C = {
  args: {
    name: "TableDemo3"
  }
}, E = {
  args: {
    name: "TableDemo4"
  }
}, F = {
  args: {
    name: "TableDemo5"
  }
}, G = {
  args: {
    name: "TableDemo6"
  }
}, H = {
  args: {
    name: "TableDemo7"
  }
}, I = {
  args: {
    name: "TableDemo8"
  }
}, J = {
  args: {
    name: "TableDemo9"
  }
}, K = {
  args: {
    name: "TableDemo10"
  }
}, L = {
  args: {
    name: "TableDemo11"
  }
}, M = {
  args: {
    name: "TableDemo12"
  }
}, O = {
  args: {
    name: "TableDemo13Virtualized"
  }
};
export {
  A as Table1,
  K as Table10,
  L as Table11,
  M as Table12,
  O as Table13Virtualized,
  B as Table2,
  C as Table3,
  E as Table4,
  F as Table5,
  G as Table6,
  H as Table7,
  I as Table8,
  J as Table9,
  w as default
};
