// @kousum/semi-ui-vue/dist/_utils/index.js
import { isVNode as l, h as m, createVNode as g, Fragment as y } from "vue";
import { cloneDeepWith as p, set as b, get as x } from "lodash";
import A from '@douyinfe/semi-foundation/lib/es/utils/warning';
import { findAll as S } from '@douyinfe/semi-foundation/lib/es/utils/getHighlight';
import { isHTMLElement as E } from '@douyinfe/semi-foundation/lib/es/utils/dom';
function k(e, n) {
  e && typeof e.stopPropagation == "function" && e.stopPropagation(), !n && e && typeof e.stopImmediatePropagation == "function" && e.stopImmediatePropagation();
}
function F(e, n) {
  return p(e, (t) => {
    if (typeof n == "function")
      return n(t);
    if (typeof t == "function" || l(t) || Object.prototype.toString.call(t) === "[object Error]")
      return t;
    if (Array.isArray(t) && t.length === 0) {
      const o = Object.keys(t);
      if (o.length) {
        const r = [];
        return o.forEach((i) => {
          b(r, i, t[i]);
        }), A(x(process, "env.NODE_ENV") !== "production", `[Semi] You may use an out-of-bounds array. In some cases, your program may not behave as expected.
                    The maximum length of an array is 4294967295.
                    Please check whether the array subscript in your data exceeds the maximum value of the JS array subscript`), r;
      } else
        return;
    }
  });
}
const I = ({
  sourceString: e = "",
  searchWords: n = [],
  option: t = {
    autoEscape: !0,
    caseSensitive: !1
  }
}) => {
  const o = S({
    sourceString: e,
    searchWords: n,
    ...t
  }), r = t.highlightTag || "mark", i = t.highlightClassName || "", a = t.highlightStyle || {};
  return o.map((f) => {
    const {
      end: d,
      start: s,
      highlight: h
    } = f, u = e.substr(s, d - s);
    return h ? m(r, {
      style: a,
      className: i
    }, u) : u;
  });
}, O = (e, {
  match: n,
  unmatch: t,
  callInInit: o = !0
}) => {
  if (typeof window < "u") {
    let i = function(a) {
      a.matches ? n && n(a) : t && t(a);
    };
    const r = window.matchMedia(e);
    return o && i(r), r.addEventListener("change", i), () => r.removeEventListener("change", i);
  }
  return null;
}, V = (e) => {
  var n, t;
  return l(e) && ((t = (n = e.type) == null ? void 0 : n.name) == null ? void 0 : t.indexOf("Icon")) > -1;
};
function w() {
  return document ? document.activeElement : null;
}
function W(e) {
  const n = w();
  return n === e || e.contains(n);
}
function v(e) {
  if (!E(e))
    return [];
  const t = ["input:not([disabled]):not([tabindex='-1'])", "textarea:not([disabled]):not([tabindex='-1'])", "button:not([disabled]):not([tabindex='-1'])", "a[href]:not([tabindex='-1'])", "select:not([disabled]):not([tabindex='-1'])", "area[href]:not([tabindex='-1'])", "iframe:not([tabindex='-1'])", "object:not([tabindex='-1'])", "*[tabindex]:not([tabindex='-1'])", "*[contenteditable]:not([tabindex='-1'])"].join(",");
  return Array.from(e.querySelectorAll(t));
}
async function N(e, n) {
  if (n === 0) {
    await e();
    return;
  } else {
    await new Promise((t) => {
      setTimeout(async () => {
        await N(e, n - 1), t();
      }, 0);
    });
    return;
  }
}
function D(e) {
  var t;
  const n = (t = e.default) == null ? void 0 : t.call(e);
  return T(n);
}
function T(e) {
  if (e && Array.isArray(e) && e.length) {
    const n = [];
    return e.forEach((t) => {
      var o;
      ((o = t.type) == null ? void 0 : o.toString()) === "Symbol(v-fgt)" && Array.isArray(t.children) ? t.children.forEach((r) => {
        n.push(r);
      }) : Array.isArray(t) ? t.map((r) => {
        n.push(r);
      }) : n.push(t);
    }), n;
  } else
    return e;
}
function c(e) {
  let n = [];
  return e.forEach((t) => {
    typeof t.type == "symbol" ? Array.isArray(t.children) && t.children.length > 0 && (n = [...n, ...c(t.children)]) : n.push(t);
  }), n;
}
function H(e) {
  return c(e);
}
function Q() {
  return globalThis && Object.prototype.toString.call(globalThis) === "[object Window]" ? window.innerWidth - document.documentElement.clientWidth : 0;
}
function q(e) {
  return isNaN(+e) ? e : e + "px";
}
function J(e) {
  return !(["Symbol(v-fgt)", "Symbol(v-cmt)", "Symbol(v-txt)"].indexOf(e.type.toString()) > -1);
}
function Y(e) {
  let n = e.default ? e.default() : null;
  if (Array.isArray(n) && n.length === 1)
    return n[0];
  if (Array.isArray(n) && n.length > 1)
    return g(y, null, [n]);
}
export {
  F as cloneDeep,
  w as getActiveElement,
  T as getChildren,
  v as getFocusableElements,
  D as getFragmentChildren,
  I as getHighLightTextHTML,
  Y as getMultinodeToFragment,
  Q as getScrollbarWidth,
  H as getVNodeChildren,
  W as isNodeContainsFocus,
  V as isSemiIcon,
  J as isVNodeTypeNotSymbol,
  O as registerMediaQuery,
  N as runAfterTicks,
  k as stopPropagation,
  q as styleNum
};
