// @kousum/semi-ui-vue/dist/space/index.js
import { defineComponent as $, createVNode as G, mergeProps as I } from "vue";
import A from "classnames";
import { cssClasses as h, strings as r } from '@douyinfe/semi-foundation/lib/es/space/constants';
import '@douyinfe/semi-foundation/lib/es/space/space.css';
import { isString as o, isNumber as n, isArray as C } from "lodash";
import { flatten as P } from "./utils";
import v from '@douyinfe/semi-foundation/lib/es/utils/getDataAttr';
const t = h.PREFIX, x = {
  wrap: {
    type: Boolean,
    default: !1
  },
  align: {
    type: String,
    default: "center"
  },
  vertical: {
    type: Boolean,
    default: !1
  },
  spacing: {
    type: [Number, Array, String],
    default: "tight"
  },
  style: [String, Object],
  className: String
}, M = /* @__PURE__ */ $({
  props: {
    ...x
  },
  name: "Space",
  setup(p, {
    slots: c
  }) {
    return () => {
      const {
        style: u,
        className: g,
        spacing: e,
        wrap: m,
        align: f,
        vertical: s
      } = p, d = m && s ? !1 : m, a = {
        ...u
      };
      let i = "", l = "";
      o(e) ? (i = e, l = e) : n(e) ? (a.rowGap = e + "px", a.columnGap = e + "px") : C(e) && (o(e[0]) ? i = e[0] : n(e[0]) && (a.columnGap = `${e[0]}px`), o(e[1]) ? l = e[1] : n(e[1]) && (a.rowGap = `${e[1]}px`));
      const S = A(t, g, {
        [`${t}-align-${f}`]: f,
        [`${t}-vertical`]: s,
        [`${t}-horizontal`]: !s,
        [`${t}-wrap`]: d,
        [`${t}-tight-horizontal`]: i === r.SPACING_TIGHT,
        [`${t}-tight-vertical`]: l === r.SPACING_TIGHT,
        [`${t}-medium-horizontal`]: i === r.SPACING_MEDIUM,
        [`${t}-medium-vertical`]: l === r.SPACING_MEDIUM,
        [`${t}-loose-horizontal`]: i === r.SPACING_LOOSE,
        [`${t}-loose-vertical`]: l === r.SPACING_LOOSE
      }), y = c.default ? P(c.default()) : null, N = v(p);
      return G("div", I(N, {
        class: S,
        style: a,
        "x-semi-prop": "children"
      }), [y]);
    };
  }
});
export {
  M as default,
  x as vuePropsType
};
