// @kousum/dnd-kit-vue/index.cjs
'use strict';

var dom = require('@dnd-kit/dom');
var hooks = require('@kousum/dnd-kit-vue/hooks');
var vue = require('vue');
var state = require('@dnd-kit/state');
var utilities = require('@kousum/dnd-kit-vue/utilities');

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
function useConfigContext() {
  const context = vue.inject("ConfigContext", vue.shallowRef(defaultManager));
  return {
    context
  };
}
var Consumer = vue.defineComponent({
  name: "ConfigContextConsumer",
  setup() {
    const slots = vue.useSlots();
    const { context } = useConfigContext();
    return () => slots.default ? slots.default(context) : null;
  }
});
var vuePropsType = {
  value: Object
};
var Provider = vue.defineComponent({
  props: __spreadValues({}, vuePropsType),
  name: "ConfigProviderProvider",
  setup(props, { slots }) {
    const ConfigContext = vue.shallowRef(props.value);
    vue.watch(
      () => props.value,
      () => {
        ConfigContext.value = props.value;
      },
      { deep: true }
    );
    vue.provide("ConfigContext", ConfigContext);
    return () => slots.default ? slots.default() : null;
  }
});
var defaultManager = new dom.DragDropManager();
var DragDropContext = {
  Provider,
  Consumer
};
function useRenderer() {
  const rendering = vue.shallowRef();
  vue.ref(0);
  const resolver = vue.shallowRef();
  const renderer = hooks.useConstant(() => ({
    get rendering() {
      var _a;
      return (_a = rendering.value) != null ? _a : Promise.resolve();
    }
  }));
  function onResolve() {
    var _a;
    (_a = resolver.value) == null ? void 0 : _a.call(resolver);
    rendering.value = void 0;
  }
  return {
    onResolve,
    renderer,
    trackRendering(callback) {
      if (!rendering.value) {
        rendering.value = new Promise((resolve) => {
          resolver.value = resolve;
        });
      }
      callback();
      onResolve();
    }
  };
}
var DragDropProvider = vue.defineComponent({
  props: {
    onCollision: Function,
    onBeforeDragStart: Function,
    onDragStart: Function,
    onDragMove: Function,
    onDragOver: Function,
    onDragEnd: Function,
    plugins: Object,
    modifiers: Object,
    sensors: Object,
    manager: Object,
    renderer: Object
    // actions: Object as PropType<any>,
    // collisionObserver: Object as PropType<any>,
    // dragOperation: Object as PropType<any>,
    // monitor: Object as PropType<any>,
    // registry: Object as PropType<any>,
  },
  name: "Test",
  setup(props, { attrs }) {
    var _a;
    const slots = vue.useSlots();
    const { renderer, trackRendering, onResolve } = useRenderer();
    const manager = vue.shallowRef(
      (_a = props.manager) != null ? _a : null
    );
    function setManager(v) {
      manager.value = v;
    }
    hooks.useEffect(() => {
      var _b;
      const _a2 = props, input = __objRest(_a2, [
        "onCollision",
        "onBeforeDragStart",
        "onDragStart",
        "onDragMove",
        "onDragOver",
        "onDragEnd"
      ]);
      const handleBeforeDragStart = props.onBeforeDragStart;
      const handleDragStart = props.onDragStart;
      const handleDragOver = props.onDragOver;
      const handleDragMove = props.onDragMove;
      const handleDragEnd = props.onDragEnd;
      const handleCollision = props.onCollision;
      const manager2 = (_b = props.manager) != null ? _b : new dom.DragDropManager(input);
      manager2.renderer = renderer.value;
      manager2.monitor.addEventListener("beforedragstart", (event, manager3) => {
        const callback = handleBeforeDragStart;
        if (callback) {
          trackRendering(() => callback(event, manager3));
        }
      });
      manager2.monitor.addEventListener(
        "dragstart",
        (event, manager3) => handleDragStart == null ? void 0 : handleDragStart(event, manager3)
      );
      manager2.monitor.addEventListener("dragover", (event, manager3) => {
        const callback = handleDragOver;
        if (callback) {
          trackRendering(() => callback(event, manager3));
        }
      });
      manager2.monitor.addEventListener("dragmove", (event, manager3) => {
        const callback = handleDragMove;
        if (callback) {
          trackRendering(() => callback(event, manager3));
        }
      });
      manager2.monitor.addEventListener("dragend", (event, manager3) => {
        const callback = handleDragEnd;
        if (callback) {
          trackRendering(() => callback(event, manager3));
        }
      });
      manager2.monitor.addEventListener(
        "collision",
        (event, manager3) => handleCollision == null ? void 0 : handleCollision(event, manager3)
      );
      setManager(manager2);
      return manager2.destroy;
    }, [renderer]);
    hooks.useOnValueChange(
      () => props.plugins,
      () => {
        var _a2;
        return manager.value && (manager.value.plugins = (_a2 = props.plugins) != null ? _a2 : dom.defaultPreset.plugins);
      }
    );
    hooks.useOnValueChange(
      () => props.sensors,
      () => {
        var _a2;
        return manager.value && (manager.value.sensors = (_a2 = props.sensors) != null ? _a2 : dom.defaultPreset.sensors);
      }
    );
    hooks.useOnValueChange(
      () => props.modifiers,
      () => {
        var _a2;
        return manager.value && (manager.value.modifiers = (_a2 = props.modifiers) != null ? _a2 : []);
      }
    );
    return () => {
      return /* @__PURE__ */ vue.h(DragDropContext.Provider, { value: manager.value }, {
        default: () => {
          var _a2;
          return (_a2 = slots.default) == null ? void 0 : _a2.call(slots);
        }
      });
    };
  }
});
function useDragDropManager() {
  return vue.inject("ConfigContext", vue.shallowRef(defaultManager));
}
function useInstance(initializer) {
  var _a;
  const manager = (_a = useDragDropManager()) != null ? _a : void 0;
  const instance = vue.shallowRef(initializer(manager.value === defaultManager ? void 0 : manager.value));
  hooks.useEffect(() => {
    instance.value.manager = manager.value;
    const unregister = instance.value.register();
    return unregister;
  }, [instance, manager]);
  return instance;
}

// src/core/draggable/useDraggable.ts
function useDraggable(input) {
  const { disabled, data, id, modifiers, sensors } = input;
  const handle = utilities.currentValue(input.handle);
  const element = utilities.currentValue(input.element);
  const draggable = useInstance(
    (manager) => new dom.Draggable(
      __spreadProps(__spreadValues({}, input), {
        handle: handle.value,
        element: element.value
      }),
      manager
    )
  );
  const isDragSource = hooks.useComputed(() => draggable.value.isDragSource);
  const status = hooks.useComputed(() => draggable.value.status);
  hooks.useOnValueChange(() => id, () => draggable.value.id = id);
  hooks.useOnValueChange(() => handle.value, () => draggable.value.handle = handle.value);
  hooks.useOnValueChange(() => element.value, () => draggable.value.element = element.value);
  hooks.useOnValueChange(() => data, () => data && (draggable.value.data = data));
  hooks.useOnValueChange(() => disabled, () => draggable.value.disabled = disabled === true);
  hooks.useOnValueChange(() => sensors, () => draggable.value.sensors = sensors);
  hooks.useOnValueChange(
    () => modifiers,
    () => draggable.value.modifiers = modifiers,
    void 0,
    state.deepEqual
  );
  hooks.useOnValueChange(
    () => input.feedback,
    () => {
      var _a;
      return draggable.value.feedback = (_a = input.feedback) != null ? _a : "default";
    }
  );
  return {
    draggable,
    isDragSource,
    status,
    handleRef: (element2) => {
      draggable.value.handle = element2 != null ? element2 : void 0;
    },
    ref: (element2) => {
      var _a, _b;
      if (!element2 && ((_a = draggable.value.element) == null ? void 0 : _a.isConnected) && !((_b = draggable.value.manager) == null ? void 0 : _b.dragOperation.status.idle)) {
        return;
      }
      draggable.value.element = element2 != null ? element2 : void 0;
    }
  };
}
function useDroppable(input) {
  const { collisionDetector, data, disabled, id, accept, type } = input;
  const element = utilities.currentValue(input.element);
  const droppable = useInstance(
    (manager) => new dom.Droppable(
      __spreadProps(__spreadValues({}, input), {
        element: element.value
      }),
      manager
    )
  );
  const isDropTarget = hooks.useComputed(() => droppable.value.isDropTarget);
  hooks.useOnValueChange(() => id, () => droppable.value.id = id);
  hooks.useOnValueChange(() => accept, () => droppable.value.id = id, void 0, state.deepEqual);
  hooks.useOnValueChange(() => collisionDetector, () => droppable.value.id = id);
  hooks.useOnValueChange(() => data, () => data && (droppable.value.data = data));
  hooks.useOnValueChange(() => disabled, () => droppable.value.disabled = disabled === true);
  hooks.useOnValueChange(() => element.value, () => droppable.value.element = element.value);
  hooks.useOnValueChange(() => type, () => droppable.value.id = id);
  return {
    isDropTarget,
    // isDropTarget: isDropTarget_,
    ref: (element2) => {
      var _a, _b;
      if (!element2 && ((_a = droppable.value.element) == null ? void 0 : _a.isConnected) && !((_b = droppable.value.manager) == null ? void 0 : _b.dragOperation.status.idle)) {
        return;
      }
      droppable.value.element = element2 != null ? element2 : void 0;
    },
    droppable
  };
}
function useDragOperation() {
  const manager = useDragDropManager();
  const source = hooks.useComputed(() => {
    var _a;
    return (_a = manager.value) == null ? void 0 : _a.dragOperation.source;
  }, [manager]);
  const target = hooks.useComputed(() => {
    var _a;
    return (_a = manager.value) == null ? void 0 : _a.dragOperation.target;
  }, [manager]);
  return {
    source,
    target
  };
}

exports.DragDropProvider = DragDropProvider;
exports.useDragDropManager = useDragDropManager;
exports.useDragOperation = useDragOperation;
exports.useDraggable = useDraggable;
exports.useDroppable = useDroppable;
exports.useInstance = useInstance;
