// @kousum/semi-ui-vue/dist/highlight/index.js
import h from "classnames";
import * as t from "../PropTypes";
import { vuePropsMake as c } from "../PropTypes";
import { cssClasses as l } from '@douyinfe/semi-foundation/lib/es/highlight/constants';
import { getHighLightTextHTML as m } from "../_utils/index";
import '@douyinfe/semi-foundation/lib/es/highlight/highlight.css';
import { defineComponent as u, useSlots as f } from "vue";
const S = l.PREFIX, C = {
  // style: PropTypes.object,
  // className: PropTypes.string,
  autoEscape: t.bool,
  caseSensitive: t.bool,
  sourceString: t.string,
  searchWords: t.array,
  highlightStyle: t.object,
  highlightClassName: t.string,
  component: t.string
}, y = {
  component: "mark",
  autoEscape: !0,
  caseSensitive: !1,
  sourceString: ""
}, T = c(C, y), N = /* @__PURE__ */ u({
  props: {
    ...T
  },
  name: "Highlight",
  setup(o, {}) {
    return f(), () => {
      const {
        searchWords: s,
        sourceString: e,
        component: r,
        highlightClassName: i,
        highlightStyle: a,
        caseSensitive: n,
        autoEscape: p
      } = o, g = h({
        [`${S}-tag`]: !0
      }, i);
      return m({
        sourceString: e,
        searchWords: s,
        option: {
          highlightTag: r,
          highlightClassName: g,
          highlightStyle: a,
          caseSensitive: n,
          autoEscape: p
        }
      });
    };
  }
});
export {
  N as default,
  T as vuePropsType
};
