// @kousum/semi-ui-vue/dist/dropdown/index.js
import { defineComponent as Z, reactive as q, ref as J, createVNode as s, mergeProps as w, cloneVNode as Q, isVNode as S } from "vue";
import C from "classnames";
import * as e from "../PropTypes";
import { vuePropsMake as W } from "../PropTypes";
import { strings as D, cssClasses as ee, numbers as I } from '@douyinfe/semi-foundation/lib/es/dropdown/constants';
import { useHasInProps as oe, useBaseComponent as ne } from "../_base/baseComponent";
import te, { vuePropsType as re } from "../tooltip";
import { numbers as se } from '@douyinfe/semi-foundation/lib/es/tooltip/constants';
import ie from '@douyinfe/semi-foundation/lib/es/dropdown/foundation';
import O from "./dropdownMenu";
import { default as Me } from "./dropdownMenu";
import k from "./dropdownItem";
import { default as ze } from "./dropdownItem";
import A from "./dropdownDivider";
import { default as Be } from "./dropdownDivider";
import _ from "./dropdownTitle";
import { default as Xe } from "./dropdownTitle";
import '@douyinfe/semi-foundation/lib/es/dropdown/dropdown.css';
import { noop as N, get as E } from "lodash";
import { useDropdownContext as le } from "./context/Consumer";
import { DropdownContext as F } from "./context";
function ae(n) {
  return typeof n == "function" || Object.prototype.toString.call(n) === "[object Object]" && !S(n);
}
D.POSITION_SET;
D.TRIGGER_SET;
const j = {
  ...re,
  onFilter: e.func,
  onFilterDropdownVisibleChange: e.func,
  onSelect: e.func,
  onHeaderCell: e.func,
  onGroupedRow: e.func,
  render: [e.node, e.func, e.object],
  visible: e.bool,
  position: e.string,
  getPopupContainer: e.func,
  mouseEnterDelay: e.number,
  mouseLeaveDelay: e.number,
  trigger: e.string,
  zIndex: e.number,
  motion: e.oneOfType([e.bool, e.func, e.object]),
  className: e.string,
  contentClassName: e.oneOfType([e.string, e.array]),
  style: e.object,
  onVisibleChange: e.func,
  rePosKey: e.oneOfType([e.string, e.number]),
  showTick: e.bool,
  prefixCls: e.string,
  spacing: e.oneOfType([e.number, e.object]),
  menu: e.array,
  name: String
}, Ke = j, ce = {
  onVisibleChange: N,
  prefixCls: ee.PREFIX,
  zIndex: se.DEFAULT_Z_INDEX,
  motion: !0,
  trigger: "hover",
  position: "bottom",
  mouseLeaveDelay: D.DEFAULT_LEAVE_DELAY,
  showTick: !1,
  closeOnEsc: !0,
  onEscKeyDown: N
}, ue = W(j, ce), pe = /* @__PURE__ */ Z({
  props: {
    ...ue
  },
  name: "Dropdown",
  setup(n, {
    slots: v,
    expose: G
  }) {
    const {
      getProps: K
    } = oe(), g = q({
      popVisible: n.visible
    }), T = J(), {
      adapter: L
    } = ne(n, g), R = function() {
      return {
        ...L(),
        setPopVisible: (o) => g.popVisible = o,
        notifyVisibleChange: (o) => {
          var i;
          return (i = n.onVisibleChange) == null ? void 0 : i.call(n, o);
        },
        getPopupId: () => T.value.getPopupId()
      };
    }(), h = new ie(R), M = (o) => {
      h.handleVisibleChange(o);
    }, {
      context: y
    } = le(), {
      level: $ = 0
    } = y.value;
    function z() {
      const {
        render: o,
        menu: i,
        contentClassName: l,
        style: c,
        showTick: t,
        prefixCls: r,
        trigger: a
      } = n, p = C(r, l), d = {
        showTick: t,
        level: $ + 1,
        trigger: a
      };
      let m = null;
      return S(o) ? m = o : Array.isArray(i) && (m = H()), s(F.Provider, {
        value: d
      }, {
        default: () => [s("div", {
          class: p,
          style: c
        }, [s("div", {
          class: `${r}-content`,
          "x-semi-prop": "render"
        }, [m])])]
      });
    }
    function H() {
      const {
        menu: o
      } = n, i = o.map((l, c) => {
        switch (l.node) {
          case "title": {
            const {
              name: t,
              node: r,
              ...a
            } = l;
            return s(_, w(a, {
              key: r + t + c
            }), {
              default: () => t
            });
          }
          case "item": {
            const {
              node: t,
              name: r,
              ...a
            } = l;
            return s(k, w(a, {
              key: t + r + c
            }), {
              default: () => r
            });
          }
          case "divider":
            return s(A, {
              key: l.node + c
            }, null);
          default:
            return null;
        }
      });
      return s(O, null, {
        default: () => i
      });
    }
    function B() {
      const {
        render: o,
        contentClassName: i,
        style: l,
        showTick: c,
        prefixCls: t
      } = n, r = C(t, i), {
        level: a = 0
      } = y.value, p = {
        showTick: c,
        level: a + 1
      };
      return s(F.Provider, {
        value: p
      }, {
        default: () => [s("div", {
          class: r,
          style: l
        }, [s("div", {
          class: `${t}-content`
        }, [o])])]
      });
    }
    return G({
      renderPopCard: B
    }), () => {
      let o;
      const {
        position: i,
        trigger: l,
        onVisibleChange: c,
        zIndex: t,
        className: r,
        motion: a,
        style: p,
        prefixCls: d,
        render: m,
        menu: me,
        showTick: fe,
        //
        onFilter: ge,
        onFilterDropdownVisibleChange: ye,
        onSelect: be,
        onHeaderCell: we,
        onGroupedRow: Ce,
        name: De,
        contentClassName: ve,
        ...U
      } = K(n);
      let {
        spacing: u
      } = n;
      const {
        level: X
      } = y.value, {
        popVisible: P
      } = g, Y = z();
      X > 0 ? u = typeof u == "number" ? u : I.NESTED_SPACING : (u === null || typeof u > "u") && (u = I.SPACING);
      let b = v.default ? v.default()[0] : null;
      return s(te, w({
        zIndex: t,
        motion: a,
        content: Y,
        className: r,
        prefixCls: d,
        spacing: u,
        position: i,
        trigger: l,
        onVisibleChange: M,
        showArrow: !1,
        returnFocusOnClose: !0,
        ref: T
      }, U), ae(o = Q(b, {
        className: C(E(b, "props.class"), {
          [`${d}-showing`]: P
        }),
        "aria-haspopup": !0,
        "aria-expanded": P,
        onKeyDown: (V) => {
          h.handleKeyDown(V);
          const x = E(b, "props.onKeyDown");
          x && x(V);
        }
      })) ? o : {
        default: () => [o]
      });
    };
  }
}), f = pe;
f.Menu = O;
f.Item = k;
f.Divider = A;
f.Title = _;
export {
  pe as Dropdown,
  Be as DropdownDivider,
  ze as DropdownItem,
  Me as DropdownMenu,
  Xe as DropdownTitle,
  Ke as DropdownVuePropsType,
  f as default,
  ue as vuePropsType
};
