// @kousum/semi-ui-vue/dist/upload/fileCard.js
import { defineComponent as oe, useSlots as se, reactive as ce, createVNode as e, mergeProps as z, Fragment as H, isVNode as ae } from "vue";
import O from "classnames";
import * as t from "../PropTypes";
import { vuePropsMake as ue } from "../PropTypes";
import { cssClasses as de, strings as o } from '@douyinfe/semi-foundation/lib/es/upload/constants';
import pe from '@douyinfe/semi-foundation/lib/es/upload/fileCardFoundation';
import { getFileSize as fe } from '@douyinfe/semi-foundation/lib/es/upload/utils';
import { IconRefresh as me, IconClear as we, IconFile as q, IconClose as he, IconAlertCircle as ge } from "@kousum/semi-icons-vue";
import B from "../locale/localeConsumer";
import Z from "../button/index";
import G from "../progress/index";
import D from "../tooltip/index";
import W from "../spin/index";
import { isElement as ve } from "../_base/reactUtils";
import { styleNum as _ } from "../_utils";
import { useBaseComponent as ye } from "../_base/baseComponent";
function Ce(r) {
  return typeof r == "function" || Object.prototype.toString.call(r) === "[object Object]" && !ae(r);
}
const n = de.PREFIX, Se = (r = {}) => (
  // @ts-ignore
  e("svg", z({
    focusable: !1,
    "aria-hidden": !0,
    width: "16",
    height: "16",
    viewBox: "0 0 16 16",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, r), [e("circle", {
    cx: "7.99992",
    cy: "7.99992",
    r: "6.66667",
    fill: "white"
  }, null), e("path", {
    "fill-rule": "evenodd",
    "clip-rule": "evenodd",
    d: "M15.3332 8.00008C15.3332 12.0502 12.0499 15.3334 7.99984 15.3334C3.94975 15.3334 0.666504 12.0502 0.666504 8.00008C0.666504 3.94999 3.94975 0.666748 7.99984 0.666748C12.0499 0.666748 15.3332 3.94999 15.3332 8.00008ZM8.99984 11.6667C8.99984 11.1145 8.55212 10.6667 7.99984 10.6667C7.44755 10.6667 6.99984 11.1145 6.99984 11.6667C6.99984 12.219 7.44755 12.6667 7.99984 12.6667C8.55212 12.6667 8.99984 12.219 8.99984 11.6667ZM7.99984 3.33341C7.27573 3.33341 6.7003 3.94171 6.74046 4.66469L6.94437 8.33495C6.97549 8.89513 7.4388 9.33341 7.99984 9.33341C8.56087 9.33341 9.02419 8.89513 9.05531 8.33495L9.25921 4.66469C9.29938 3.94171 8.72394 3.33341 7.99984 3.33341Z",
    fill: "#F93920"
  }, null)])
), be = (r = {}) => (
  // @ts-ignore
  e("svg", z({
    focusable: !1,
    "aria-hidden": !0,
    width: "28",
    height: "28",
    viewBox: "0 0 28 28",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, r), [e("circle", {
    cx: "14",
    cy: "14",
    r: "14",
    fill: "#16161A",
    "fill-opacity": "0.6"
  }, null), e("path", {
    d: "M9 10.25V18.25L10.25 13.25H17.875V11.75C17.875 11.4739 17.6511 11.25 17.375 11.25H14L12.75 9.75H9.5C9.22386 9.75 9 9.97386 9 10.25Z",
    stroke: "currentColor",
    "stroke-linecap": "round",
    "stroke-linejoin": "round"
  }, null), e("path", {
    d: "M18 18.25L19 13.25H10.2031L9 18.25H18Z",
    stroke: "currentColor",
    "stroke-linecap": "round",
    "stroke-linejoin": "round"
  }, null)])
), Ie = (r = {}) => (
  // @ts-ignore
  e("svg", z({
    focusable: !1,
    "aria-hidden": !0,
    width: "24",
    height: "24",
    viewBox: "0 0 24 24",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, r), [e("path", {
    d: "M6 17V7.58824C6 7.26336 6.26863 7 6.6 7H10.5L12 8.76471H16.05C16.3814 8.76471 16.65 9.02806 16.65 9.35294V11.1176H7.5L6 17ZM6 17L7.44375 11.1176H18L16.8 17L6 17Z",
    stroke: "currentColor",
    "stroke-width": "1.5",
    "stroke-linecap": "round",
    "stroke-linejoin": "round"
  }, null)])
), Le = {
  className: t.string,
  disabled: {
    type: t.bool,
    required: !0
  },
  listType: {
    type: t.string,
    required: !0
  },
  name: {
    type: t.string,
    required: !0
  },
  onPreviewClick: {
    type: t.func,
    required: !0
  },
  onRemove: {
    type: t.func,
    required: !0
  },
  onReplace: {
    type: t.func,
    required: !0
  },
  onRetry: {
    type: t.func,
    required: !0
  },
  percent: t.number,
  preview: t.bool,
  previewFile: t.func,
  picWidth: t.oneOfType([t.number, t.string]),
  picHeight: t.oneOfType([t.number, t.string]),
  showReplace: t.bool,
  showRetry: t.bool,
  size: {
    type: t.oneOfType([t.number, t.string]),
    required: !0
  },
  status: {
    type: t.string,
    required: !0
  },
  style: t.object,
  url: t.string,
  validateMessage: t.node,
  index: t.number,
  // key: String,
  showPicInfo: Boolean,
  renderPicInfo: Function,
  renderPicPreviewIcon: Function,
  renderFileOperation: Function,
  uid: {
    type: String,
    required: !0
  },
  fileInstance: Object,
  renderThumbnail: Function,
  response: Object,
  event: Object,
  shouldUpload: Boolean,
  _sizeInvalid: Boolean
}, Te = {
  listType: o.FILE_LIST_DEFAULT,
  name: "",
  onRemove: () => {
  },
  onRetry: () => {
  },
  preview: !1,
  size: ""
}, Fe = ue(Le, Te), je = /* @__PURE__ */ oe({
  props: {
    ...Fe
  },
  name: "FileCard",
  setup(r, {}) {
    se();
    const y = ce({
      fallbackPreview: !1
    }), {
      adapter: X
    } = ye(r, y);
    function J() {
      return {
        ...X(),
        updateFallbackPreview: (l) => {
          y.fallbackPreview = l;
        }
      };
    }
    const K = J(), M = new pe(K);
    function Q(l) {
      return typeof l == "number" ? fe(l) : l;
    }
    function Y() {
      const {
        status: l,
        validateMessage: i
      } = r;
      let s = null;
      switch (!0) {
        case (typeof i == "string" && l === o.FILE_STATUS_VALIDATING):
          s = e(H, null, [e(W, {
            size: "small",
            wrapperClassName: `${n}-file-card-icon-loading`
          }, null), i]);
          break;
        case typeof i == "string":
          s = e(H, null, [e(ge, {
            class: `${n}-file-card-icon-error`
          }, null), i]);
          break;
        case ve(i):
          s = i;
          break;
      }
      return s;
    }
    function ee() {
      const {
        status: l,
        validateMessage: i
      } = r;
      let s = null;
      switch (!0) {
        case (i && l === o.FILE_STATUS_VALIDATING):
          s = e(W, {
            size: "small",
            wrapperClassName: `${n}-picture-file-card-icon-loading`
          }, null);
          break;
        case (i && (l === o.FILE_STATUS_VALID_FAIL || l === o.FILE_STATUS_UPLOAD_FAIL)):
          s = e("div", {
            class: `${n}-picture-file-card-icon-error`
          }, [e(Se, null, null)]);
          break;
      }
      return s ? e(D, {
        content: i,
        trigger: "hover",
        position: "bottom"
      }, Ce(s) ? s : {
        default: () => [s]
      }) : null;
    }
    function re(l) {
      const {
        fallbackPreview: i
      } = y, {
        url: s,
        percent: f,
        status: d,
        disabled: C,
        style: S,
        onPreviewClick: h,
        showPicInfo: g,
        renderPicInfo: u,
        renderPicPreviewIcon: b,
        renderThumbnail: v,
        name: I,
        index: L,
        picHeight: m,
        picWidth: w
      } = r, c = d === o.FILE_STATUS_UPLOADING && f !== 100, A = d === o.FILE_STATUS_UPLOAD_FAIL && r.showRetry, P = d === o.FILE_STATUS_SUCCESS && r.showReplace, k = d === o.FILE_STATUS_SUCCESS && !r.showReplace, T = typeof v == "function", $ = O({
        [`${n}-picture-file-card`]: !0,
        [`${n}-picture-file-card-preview-fallback`]: i,
        [`${n}-picture-file-card-disabled`]: C,
        [`${n}-picture-file-card-show-pointer`]: typeof h < "u",
        [`${n}-picture-file-card-error`]: d === o.FILE_STATUS_UPLOAD_FAIL,
        [`${n}-picture-file-card-uploading`]: c,
        [`${n}-picture-file-card-custom-thumbnail`]: T && m && w
      }), E = e("div", {
        role: "button",
        "tab-index": 0,
        class: `${n}-picture-file-card-retry`,
        onClick: (p) => j(p)
      }, [e(me, {
        class: `${n}-picture-file-card-icon-retry`
      }, null)]), F = e(D, {
        trigger: "hover",
        position: "top",
        content: l.replace,
        showArrow: !1,
        spacing: 4
      }, {
        default: () => [e("div", {
          role: "button",
          "tab-index": 0,
          class: `${n}-picture-file-card-replace`,
          onClick: (p) => V(p)
        }, [e(be, {
          class: `${n}-picture-file-card-icon-replace`
        }, null)])]
      }), U = e("div", {
        class: `${n}-picture-file-card-preview`
      }, [typeof b == "function" ? b(r) : null]), a = e("div", {
        role: "button",
        "tab-index": 0,
        class: `${n}-picture-file-card-close`,
        onClick: (p) => N(p)
      }, [e(we, {
        class: `${n}-picture-file-card-icon-close`
      }, null)]), ne = typeof u == "function" ? u(r) : e("div", {
        class: `${n}-picture-file-card-pic-info`
      }, [L + 1]);
      let R = {}, x = S ? {
        ...S
      } : {};
      m && (x.height = _(m), R.height = _(m)), w && (x.width = _(w), R.width = _(w));
      const le = i ? e(q, {
        size: "large"
      }, null) : e("img", {
        src: s,
        alt: I,
        onError: (p) => M.handleImageError(p),
        style: R
      }, null), ie = T ? v(r) : le;
      return e("div", {
        role: "listitem",
        class: $,
        style: x,
        onClick: h
      }, [ie, c ? e(G, {
        percent: f,
        type: "circle",
        size: "small",
        orbitStroke: "#FFF",
        "aria-label": "uploading file progress"
      }, null) : null, A ? E : null, P && F, k && U, g && ne, !C && a, ee()]);
    }
    function te(l) {
      const {
        name: i,
        size: s,
        percent: f,
        url: d,
        showRetry: C,
        showReplace: S,
        preview: h,
        previewFile: g,
        status: u,
        style: b,
        onPreviewClick: v,
        renderFileOperation: I
      } = r, {
        fallbackPreview: L
      } = y, m = O({
        [`${n}-file-card`]: !0,
        [`${n}-file-card-fail`]: u === o.FILE_STATUS_VALID_FAIL || u === o.FILE_STATUS_UPLOAD_FAIL,
        [`${n}-file-card-show-pointer`]: typeof v < "u"
      }), w = O({
        [`${n}-file-card-preview`]: !0,
        [`${n}-file-card-preview-placeholder`]: !h || g || L
      }), c = `${n}-file-card-info`, A = `${n}-file-card-close`, P = `${n}-file-card-replace`, k = !(f === 100 || typeof f > "u") && u === o.FILE_STATUS_UPLOADING, T = u === o.FILE_STATUS_UPLOAD_FAIL && C, $ = u === o.FILE_STATUS_SUCCESS && S, E = Q(s);
      let F = h && !L ? e("img", {
        src: d,
        alt: i,
        onError: (a) => M.handleImageError(a)
      }, null) : e(q, {
        size: "large"
      }, null);
      g && (F = g(r));
      const U = typeof I == "function" ? I(r) : e(Z, {
        onClick: (a) => N(a),
        type: "tertiary",
        icon: e(he, null, null),
        theme: "borderless",
        size: "small",
        class: A
      }, null);
      return e("div", {
        role: "listitem",
        class: m,
        style: b,
        onClick: v
      }, [e("div", {
        class: w
      }, [F]), e("div", {
        class: `${c}-main`
      }, [e("div", {
        class: `${c}-main-text`
      }, [e("span", {
        class: `${c}-name`
      }, [i]), e("span", null, [e("span", {
        class: `${c}-size`
      }, [E]), $ && e(D, {
        trigger: "hover",
        position: "top",
        showArrow: !1,
        content: l.replace
      }, {
        default: () => [e(Z, {
          onClick: (a) => V(a),
          type: "tertiary",
          theme: "borderless",
          size: "small",
          icon: e(Ie, null, null),
          className: P
        }, null)]
      })])]), k ? e(G, {
        percent: f,
        style: {
          width: "100%"
        },
        "aria-label": "uploading file progress"
      }, null) : null, e("div", {
        class: `${c}-main-control`
      }, [e("span", {
        class: `${c}-validate-message`
      }, [Y()]), T ? e("span", {
        role: "button",
        "tab-index": 0,
        class: `${c}-retry`,
        onClick: (a) => j(a)
      }, [l.retry]) : null])]), U]);
    }
    function N(l) {
      l.stopPropagation(), r.onRemove();
    }
    function V(l) {
      l.stopPropagation(), r.onReplace();
    }
    function j(l) {
      l.stopPropagation(), r.onRetry();
    }
    return () => {
      const {
        listType: l
      } = r;
      return l === o.FILE_LIST_PIC ? e(B, {
        componentName: "Upload"
      }, {
        default: (i) => re(i)
      }) : l === o.FILE_LIST_DEFAULT ? e(B, {
        componentName: "Upload"
      }, {
        default: (i) => te(i)
      }) : null;
    };
  }
});
export {
  je as default,
  Fe as vuePropsType
};
