// @kousum/semi-ui-vue/dist/backtop/index.js
import { defineComponent as v, useSlots as w, reactive as T, onMounted as B, onBeforeUnmount as P, createVNode as s, mergeProps as I } from "vue";
import x from "classnames";
import * as o from "../PropTypes";
import { vuePropsMake as F } from "../PropTypes";
import { throttle as H } from "lodash";
import { cssClasses as U } from '@douyinfe/semi-foundation/lib/es/backtop/constants';
import j from '@douyinfe/semi-foundation/lib/es/backtop/foundation';
import '@douyinfe/semi-foundation/lib/es/backtop/backtop.css';
import D from "../iconButton";
import { IconChevronUp as E } from "@kousum/semi-icons-vue";
import { useBaseComponent as M } from "../_base/baseComponent";
import { getFragmentChildren as N } from "../_utils";
const S = U.PREFIX, V = () => window, l = {
  visibilityHeight: 400,
  target: V,
  duration: 450
}, W = {
  target: o.func,
  visibilityHeight: o.number,
  duration: o.number,
  onClick: o.func,
  style: o.object,
  className: o.string
}, R = F(W, l), ot = /* @__PURE__ */ v({
  props: {
    ...R
  },
  name: "BackTop",
  setup(e, {}) {
    const a = w();
    let c;
    const i = T({
      visible: !1
    }), {
      adapter: m
    } = M(e, i);
    function u() {
      return {
        ...m(),
        updateVisible: (t) => {
          i.visible = t;
        },
        notifyClick: (t) => {
          e.onClick && e.onClick(t);
        },
        targetIsWindow: (t) => t === window,
        isWindowUndefined: () => window === void 0,
        targetScrollToTop: (t, n) => {
          t === window ? (document.body.scrollTop = n, document.documentElement.scrollTop = n) : t.scrollTop = n;
        }
      };
    }
    const d = u(), r = new j(d);
    B(() => {
      r.init(), c = H(p, e.duration ?? l.duration);
    }), P(() => {
      r.destroy();
    });
    function p(t) {
      r.onClick(t);
    }
    function f() {
      return s(D, {
        theme: "light",
        icon: s(E, null, null)
      }, null);
    }
    return () => {
      const t = N(a), {
        className: n,
        style: C,
        onClick: X,
        visibilityHeight: _,
        target: q,
        ...g
      } = e, {
        visible: h
      } = i, k = x(S, n), b = t || f();
      return h ? (
        // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
        s("div", I(g, {
          class: k,
          style: C,
          onClick: (y) => c(y),
          "x-semi-prop": "children"
        }), [b])
      ) : null;
    };
  }
});
export {
  ot as default,
  R as vuePropsType
};
