// @kousum/semi-ui-vue/dist/form/field.js
import o from "./hoc/withField";
import r from "../input/index";
import t from "../input/textArea";
import m from "../inputNumber/index";
import n from "../select/index";
import i from "../select/option";
import c from "../select/optionGroup";
import { Checkbox as a } from "../checkbox/index";
import u from "../checkbox/checkboxGroup";
import { Radio as p } from "../radio/index";
import d from "../radio/radioGroup";
import f from "../datePicker/index";
import l from "../switch/index";
import s from "../slider/index";
import F from "../timePicker/index";
import h from "../treeSelect/index";
import C from "../cascader/index";
import S from "../rating/index";
import k from "../autoComplete/index";
import g from "../upload/index";
import v from "../tagInput/index";
const J = o(r, {
  maintainCursor: !0
}), M = o(m, {
  maintainCursor: !0
}), Q = o(t, {
  maintainCursor: !0
}), e = o(n, void 0), I = i;
e.Option = I;
const V = e, W = c, X = o(u, void 0), Y = o(a, {
  valueKey: "checked",
  valuePath: "target.checked",
  shouldInject: !1
}), Z = o(d, {
  valuePath: "target.value"
}), _ = o(p, {
  valueKey: "checked",
  valuePath: "target.checked",
  shouldInject: !1
}), $ = o(f, void 0), oo = o(l, {
  valueKey: "checked"
}), eo = o(s, void 0), ro = o(F, void 0), to = o(h, void 0), mo = o(C, void 0), no = o(S, void 0), io = o(k, {
  valueKey: "value",
  onKeyChangeFnName: "onChange"
}), co = o(g, {
  valueKey: "fileList",
  valuePath: "fileList",
  onKeyChangeFnName: "onChange"
}), ao = o(v, void 0);
export {
  io as FormAutoComplete,
  mo as FormCascader,
  Y as FormCheckbox,
  X as FormCheckboxGroup,
  $ as FormDatePicker,
  J as FormInput,
  M as FormInputNumber,
  _ as FormRadio,
  Z as FormRadioGroup,
  no as FormRating,
  V as FormSelect,
  I as FormSelectOption,
  W as FormSelectOptionGroup,
  eo as FormSlider,
  oo as FormSwitch,
  ao as FormTagInput,
  Q as FormTextArea,
  ro as FormTimePicker,
  to as FormTreeSelect,
  co as FormUpload
};
