// @kousum/semi-ui-vue/dist/form/label.js
import { defineComponent as P, createVNode as t, Fragment as T } from "vue";
import q from "classnames";
import * as e from "../PropTypes";
import { vuePropsMake as w } from "../PropTypes";
import { cssClasses as F } from '@douyinfe/semi-foundation/lib/es/form/constants';
import { LocaleConsumerFunc as L } from "../locale/localeConsumer";
import { styleNum as v } from "../_utils";
const E = L(), l = F.PREFIX, V = {
  id: e.string,
  required: e.bool,
  text: e.node,
  disabled: e.bool,
  name: e.string,
  align: e.string,
  width: e.oneOfType([e.number, e.string]),
  style: e.object,
  className: e.string,
  extra: e.node,
  optional: e.bool
}, j = {
  required: !1,
  name: "",
  align: "left",
  className: "",
  optional: !1
}, k = w(V, j), X = /* @__PURE__ */ P({
  props: k,
  name: "Label",
  setup(f, {
    slots: a
  }) {
    return () => {
      const m = a.default ? a.default() : null, {
        required: c,
        text: s,
        disabled: u,
        name: b,
        width: n,
        align: r,
        style: i,
        className: x,
        extra: o,
        id: g,
        optional: y
      } = f, $ = q(x, {
        [`${l}-field-label`]: !0,
        [`${l}-field-label-left`]: r === "left",
        [`${l}-field-label-right`]: r === "right",
        [`${l}-field-label-required`]: c,
        [`${l}-field-label-disabled`]: u,
        [`${l}-field-label-with-extra`]: o
      }), d = i || {};
      n && (d.width = v(n));
      const h = t(E, {
        componentName: "Form"
      }, {
        default: (N) => t("span", {
          class: `${l}-field-label-optional-text`
        }, [N.optional])
      }), p = t("div", {
        class: `${l}-field-label-text`,
        "x-semi-prop": "label"
      }, [typeof s < "u" ? s : m, y ? h : null]), C = t(T, null, [p, t("div", {
        class: `${l}-field-label-extra`
      }, [o])]);
      return t("label", {
        class: $,
        for: b,
        style: d,
        id: g
      }, [o ? C : p]);
    };
  }
});
export {
  k as VuePropsType,
  X as default
};
