// @kousum/semi-ui-vue/dist/popover/Arrow.js
import { defineComponent as b, createVNode as o } from "vue";
import O from "classnames";
import { get as r } from "lodash";
import { cssClasses as C, strings as s, numbers as c } from '@douyinfe/semi-foundation/lib/es/popover/constants';
const u = {
  position: String,
  className: String,
  arrowStyle: [String, Object],
  popStyle: [String, Object]
}, y = /* @__PURE__ */ b({
  props: {
    ...u
  },
  name: "Arrow",
  setup(S, {
    slots: f
  }) {
    return () => {
      const {
        position: e = "",
        className: d,
        arrowStyle: t,
        popStyle: l,
        ...L
      } = S, g = e.indexOf("top") === 0 || e.indexOf("bottom") === 0, m = O(d, C.ARROW), n = r(t, "borderOpacity", s.DEFAULT_ARROW_STYLE.borderOpacity), i = r(t, "backgroundColor", r(l, "backgroundColor", s.DEFAULT_ARROW_STYLE.backgroundColor)), p = r(t, "borderColor", r(l, "borderColor", s.DEFAULT_ARROW_STYLE.borderColor)), a = {
        ...L,
        width: c.ARROW_BOUNDING.width,
        height: c.ARROW_BOUNDING.height,
        xmlns: "http://www.w3.org/2000/svg",
        class: m
      };
      return g ? o("svg", a, [o("path", {
        d: "M0 0.5L0 1.5C4 1.5, 5.5 3, 7.5 5S10,8 12,8S14.5 7, 16.5 5S20,1.5 24,1.5L24 0.5L0 0.5z",
        fill: p,
        opacity: n
      }, null), o("path", {
        d: "M0 0L0 1C4 1, 5.5 2, 7.5 4S10,7 12,7S14.5  6, 16.5 4S20,1 24,1L24 0L0 0z",
        fill: i
      }, null)]) : o("svg", a, [o("path", {
        d: "M0.5 0L1.5 0C1.5 4, 3 5.5, 5 7.5S8,10 8,12S7 14.5, 5 16.5S1.5,20 1.5,24L0.5 24L0.5 0z",
        fill: p,
        opacity: n
      }, null), o("path", {
        d: "M0 0L1 0C1 4, 2 5.5, 4 7.5S7,10 7,12S6 14.5, 4 16.5S1,20 1,24L0 24L0 0z",
        fill: i
      }, null)]);
    };
  }
});
export {
  y as default,
  u as vuePropsType
};
