// @kousum/semi-ui-vue/dist/withVModel/index.js
import o from "./WithVModel";
import e from "../input";
import t from "../autoComplete";
import r from "../cascader";
import { CheckboxGroup as m } from "../checkbox";
import i from "../datePicker";
import l from "../inputNumber";
import c from "../radio/radioGroup";
import p from "../rating";
import n from "../select";
import a from "../slider";
import d from "../switch";
import s from "../tagInput";
import f from "../timePicker";
import u from "../transfer";
import M from "../treeSelect";
import V from "../upload";
const A = o(t), D = o(r), L = o(m), U = o(i), F = o(e), W = o(l), j = o(c, { valuePath: "target.value" }), q = o(p), z = o(n), B = o(a), E = o(d, { valueKey: "checked" }), H = o(s), J = o(f), O = o(u), Q = o(M), X = o(V, {
  valueKey: "fileList",
  valuePath: "fileList",
  onKeyChangeFnName: "onChange"
});
export {
  A as AutoCompleteVModel,
  D as CascaderVModel,
  L as CheckboxGroupVModel,
  U as DatePickerVModel,
  W as InputNumberVModel,
  F as InputVModel,
  j as RadioGroupVModel,
  q as RatingVModel,
  z as SelectVModel,
  B as SliderVModel,
  E as SwitchVModel,
  H as TagInputVModel,
  J as TimePickerVModel,
  O as TransferVModel,
  Q as TreeSelectVModel,
  X as UploadVModel
};
